# Libraries
```{r}
library(plyr)
library(tidyverse)
library(pheatmap)
library(ggplot2)
library(ape)
library(ComplexHeatmap)
library(circlize)
library(edgeR)
library(readxl)
library(umap)
library(ecodist)
library(rnndescent)
library(uwot)
library(phyloseq)
library(scales)
library(randomcoloR)
library(magrittr)
library(cowplot)
library(gridExtra)
library(viridis)
library(ANCOMBC)
library(MicrobiomeProfiler)
library(ggrepel)
library(e1071)
source("BASE_DIR/SCRIPTS/omics_analyses_functions.R")
source("BASE_DIR/SCRIPTS/PLASMIDOME/plasmidome_analysis_functions.R")
source("BASE_DIR/SCRIPTS/diversity.R")
```

```{r}
# Color palettes
colors <- list(disease = c("Healthy" = "#7AD151FF", "UC" = "#FDE725FF", "CD" = "#414487FF"), 
               project = c("POP" = "#46F884FF", "IBD Validation" = "#F05B12FF", "Longitudinal Mycobiome" = "#3E9BFEFF"),
               gender = c("m" = "#395D9CFF", "f" = "#60CEACFF"), 
               circularity = c("#FCA50AFF", "#932667FF"), 
               only_complete = c("all" = "#FFEA46FF", "only_complete" = "#31446BFF"), 
               general_palette_other = c(viridis_pal(option = "H", begin = 0.3, end = 0.1, direction = 1)(3), 
                                         viridis_pal(direction = 1)(10), viridis_pal(option = "C", direction = -1)(10), 
                                         viridis_pal(option = "H", begin = 1, end = 0.77)(3), "#BEBEBE" ),
               general_palette = c(viridis_pal(option = "H", begin = 0.3, end = 0.1, direction = 1)(3), 
                                   viridis_pal(direction = 1)(10), viridis_pal(option = "C", direction = -1)(10), 
                                   viridis_pal(option = "H", begin = 1, end = 0.77)(3)),
               sensitivity = c("passed" = "black", "failed" = "#606060", "struc_zero" = "red"))

show_col(colors$general_palette)
```



# Read input files
## Metadata

```{r}

sample_list <- read.table(file = 'BASE_DIR/SAMPLE_LISTS/AllSamples.txt', header = F, sep = '\t', 
                          col.names = c("SubjectID", "Study"), colClasses = c("character", "factor")) %>% 
  mutate(Study = case_when(Study == "POP" ~ "POP", Study == "LONGITUDINAL_MYCOBIOME" ~ "Longitudinal Mycobiome",
                           Study == "LUISITO_IBD" ~ "IBD Validation"))

Luisito_metadata <- read_xlsx(path = 'BASE_DIR/SAMPLE_LISTS/metadata_IBD_Validation_CD_UC_Nov_24_GSG.xlsx',
                              sheet = 1, col_names = T) %>% 
  rename(Disease = `IBD type(1=CD, 2=UC)`) %>%
  mutate(SubjectID = gsub(pattern = '.', replacement = '_', x = paste0(SubjectID, 'B'), fixed = T)) %>% 
  filter(SubjectID %in% sample_list$SubjectID, Disease != '') %>% 
  mutate(Gender = case_when(gender == 'H' ~ 'm', gender == 'M' ~ 'f'))
# Fix some heights
Luisito_metadata[Luisito_metadata$SubjectID == "VH_CD_031B", "height (m)"] <- "1.72"
Luisito_metadata[Luisito_metadata$SubjectID == "VH_CD_030B", "height (m)"] <- "1.76"
Luisito_metadata[Luisito_metadata$SubjectID == "VH_CD_019B", "height (m)"] <- "1.5"
Luisito_metadata[Luisito_metadata$SubjectID == "VH_UC_020B", "height (m)"] <- "1.7"
Luisito_metadata[Luisito_metadata$SubjectID == "VH_UC_012B", "height (m)"] <- "1.7"
Luisito_metadata[Luisito_metadata$SubjectID == "VH_CD_039B", "height (m)"] <- "1.7"
Luisito_metadata$`height (m)` <- as.numeric(Luisito_metadata$`height (m)`)
Luisito_metadata$`weigth (kg)` <- as.numeric(Luisito_metadata$`weigth (kg)`)
Luisito_metadata <- Luisito_metadata %>% mutate(BMI_value = `weigth (kg)` / (Luisito_metadata$`height (m)`)^2 ) %>% 
  mutate(BMI = case_when(BMI_value < 18.5 ~ 'underweight', BMI_value < 25 ~ 'healthy weight',
                         BMI_value < 30 ~ 'overweight', BMI_value >= 30 ~ 'obesity')) %>% 
  mutate(Height = `height (m)` * 100) %>%
  rename(Weight = `weigth (kg)`)


Longitudinal_metadata <- read.table('/mnt/synology/ZIXUAN/LongitudinalMycobiome/metadata.csv', header = T, sep = '\t') %>%
  filter(SampleID %in% sample_list$SubjectID) %>% mutate(Gender = case_when(Gender == 'Male' ~ 'm', Gender == 'Female' ~ 'f')) %>%
  mutate(Disease = 'Healthy') %>%  
  rename(SubjectID = SampleID)
Longitudinal_metadata[Longitudinal_metadata$BMI == "healthy weight ", "BMI"] <- "healthy weight"

```


Process POP metadata 
```{r}
metadata <- read.table(file = 'BASE_DIR/WEB_FILES/Integred_met_aug24.tsv', sep = '\t', header = T)
colnames(metadata) <- gsub("\\.", "_", colnames(metadata))
diversity <- read.table(file = '/mnt/synology/DATA_ANALYSIS/POP_STUDY/alpha_div_bact.csv', sep = ',', header = T) %>%
  rename(Novogene_ID = ID)


metadata$ID <- gsub(pattern = 'Hc_D', replacement = 'HC_D', x = metadata$ID, ignore.case = F, fixed = T)
metadata$Code <- gsub(pattern = 'Hc_D', replacement = 'HC_D', x = metadata$Code, ignore.case = F, fixed = T)

## Change first sample for individual who changed gender between questionnaires
metadata[metadata$ID == 'HC_D1046_0', 'Gender'] <- 'other'

# Group by each ID and calculate mean values (only relevant variables)
# all genders must be the same for a single individual
selected_metadata <- metadata %>% rename(n_Survey = n_Survey_y, hPDI_Adj = hDPI_Adj, uPDI_Adj = uDPI_Adj)

integrated_metadata <- left_join(selected_metadata, diversity %>% select(Chao1, Shannon, Novogene_ID), by = 'Novogene_ID')

############################## MOMENTARY FIX FOR SOME PDI indexes that were calculated wrong in 2nd timepoint
PDI_fix <- read.table(file = 'BASE_DIR/WEB_FILES/dpi_adj_TP2_aug24.csv', sep = '\t', header = T)
PDI_fix$RN_ID <- gsub(pattern = 'Hc_D', replacement = 'HC_D', x = PDI_fix$RN_ID, ignore.case = F, fixed = T)

integrated_metadata[match(PDI_fix$RN_ID, integrated_metadata$ID), 'hPDI_Adj'] <- PDI_fix$hDPI
integrated_metadata[match(PDI_fix$RN_ID, integrated_metadata$ID), 'uPDI_Adj'] <- PDI_fix$uDPI

POP_metadata <- integrated_metadata %>% filter(Novogene_ID %in% sample_list$SubjectID)
```

Join everything
```{r}
# Important metadata columns: Weight, Height, Age, Gender, BMI_value, BMI, Disease, SubjectID

join_POP <- POP_metadata %>% rename(SubjectID = Novogene_ID) %>% mutate(Disease = 'Healthy', Study = 'POP') %>% 
  rename(BMI_value = BMI) %>%
  mutate(BMI = case_when(BMI_value < 18.5 ~ 'underweight', BMI_value < 25 ~ 'healthy weight',
                         BMI_value < 30 ~ 'overweight', BMI_value >= 30 ~ 'obesity')) %>%
  select(Study, SubjectID, Disease, Age, Gender, Weight, Height, BMI, BMI_value)
join_Luisito <- Luisito_metadata %>% mutate(Study = 'IBD Validation') %>%
  select(Study, SubjectID, Disease, Age, Gender, Weight, Height, BMI, BMI_value)
join_longitudinal <- Longitudinal_metadata %>% mutate(Study = 'Longitudinal Mycobiome') %>%
  select(Study, SubjectID, Disease, Age, Gender, Weight, Height, BMI, BMI_value)

joint_metadata <- left_join(sample_list, rbind(join_POP, join_Luisito, join_longitudinal), by = c('SubjectID', 'Study')) %>%
  filter(!is.na(Disease))
```


## Sequencing depth
```{r}
n_reads <- read.table(file = 'BASE_DIR/SAMPLE_LISTS/sequencing_stats.txt', 
                      header = F, sep = ' ', col.names = c('SubjectID', 'n_reads'), skip = 1)
joint_metadata <- left_join(joint_metadata, n_reads, by = 'SubjectID')
rownames(joint_metadata) <- joint_metadata$SubjectID

```

## Counts tables
```{r}
coverage <- read.table(file = 'BASE_DIR/PROFILING/coverage_only-complete.tsv', header = T, sep = '\t') %>% 
  select(all_of(c('pCode', joint_metadata$SubjectID)))
rownames(coverage) <- coverage$pCode
numreads <- read.table(file = 'BASE_DIR/PROFILING/numreads_only-complete.tsv', header = T, sep = '\t') %>%
  select(all_of(c('pCode', joint_metadata$SubjectID)))
rownames(numreads) <- numreads$pCode

# ensure a minimum of 50% coverage or else count it as missing
numreads[coverage < 50] <- 0


#meandepth <- read.table(file = 'BASE_DIR/PROFILING/meandepth.tsv', header = T, sep = '\t') %>% select(all_of(c('pCode', joint_metadata$SubjectID)))
#rownames(meandepth) <- meandepth$pCode

coverage_with_fragments <- read.table(file = 'BASE_DIR/PROFILING/coverage.tsv', 
                                     header = T, sep = '\t') %>%select(all_of(c('pCode', joint_metadata$SubjectID)))
rownames(coverage_with_fragments) <- coverage_with_fragments$pCode
numreads_with_fragments <- read.table(file = 'BASE_DIR/PROFILING/numreads.tsv', 
                                     header = T, sep = '\t') %>% select(all_of(c('pCode', joint_metadata$SubjectID)))
rownames(numreads_with_fragments) <- numreads_with_fragments$pCode

# ensure a minimum of 50% coverage or else count it as missing
numreads_with_fragments[coverage_with_fragments < 50] <- 0

#meandepth_with_fragments <- read.table(file = 'BASE_DIR/PROFILING/meandepth_only-complete.tsv', header = T, sep = '\t') %>% select(all_of(c('pCode', joint_metadata$SubjectID)))
#rownames(meandepth_with_fragments) <- meandepth_with_fragments$pCode
```

## Plasmid info 
```{r}
plasmid_clusters <- read.table("BASE_DIR/MOBMESS/plasmids-mobmess_clusters.txt", header = T, sep = '\t', 
                               na.strings = c("NA", "")) %>% filter(contig %in% rownames(numreads)) %>% rename(Plasmid_ID = contig) 

plasmid_info <- read.table(file = 'BASE_DIR/PLASMIDS/nonredundant_plasmids_info.txt', 
                           header = T, sep = '\t', na.strings = c("NA", "")) %>% 
  right_join(y = plasmid_clusters %>% select(Plasmid_ID, cluster_type, systems), by = "Plasmid_ID") 

rownames(plasmid_info) <- plasmid_info$Plasmid_ID

CARD_resistance <- read.table("BASE_DIR/CARD/CARD_hits.txt", header = T, sep = "\t", comment.char = "", quote = "") %>% mutate(Plasmid_ID = unlist(strsplit(Contig, split = "_[0-9]+$"))) %>% group_by(Plasmid_ID) %>% 
  summarise(ARG = T, n_ARG = n(), Best_Hit_ARO = paste0(Best_Hit_ARO, collapse = ";"), Drug.Class = paste0(unique(Drug.Class), collapse = ";"),
            Resistance.Mechanism = paste0(unique(Resistance.Mechanism), collapse = ";"), 
            AMR.Gene.Family = paste0(unique(AMR.Gene.Family), collapse = ";"),
            Antibiotic = paste0(unique(Antibiotic), collapse = ";"))

plasmid_info <- plasmid_info %>% left_join(CARD_resistance, by = "Plasmid_ID")

```

## Gene content
```{r}
genecalls <- read.table(file = "BASE_DIR/ANVIO/FINAL_PLASMIDS/nonredundant_plasmids-gene-calls.txt", 
                        header = T, sep = "\t", quote = "") %>% select(-aa_sequence)
cogs_pfams <- read.table(file = "BASE_DIR/ANVIO/FINAL_PLASMIDS/nonredundant_plasmids-cogs-and-pfams.txt",
                         quote = "", header = T, sep = "\t")

gene_content <- right_join(genecalls, cogs_pfams, by = "gene_callers_id") %>% rename(Plasmid_ID = contig)
```


# Phyloseq object

```{r}
MET <- sample_data(joint_metadata)
rownames(MET) <- rownames(joint_metadata)

# Get a list with plasmids and their plasmid systems as a taxonomy approximation 
# However, some plasmids are assigned to multiple systems
# Instead, cluster all systems containing overlapping plasmids into just one
plasmid_systems <- plasmid_clusters %>% select(Plasmid_ID, systems)
# Get a list of all systems
systems_list <- data.frame(system = plasmid_systems %>% filter(!is.na(systems)) %>% pull(systems) %>% 
                             strsplit(split = "|", fixed = T) %>% unlist() %>% unique(),
                           representative = as.character(NA))
rownames(systems_list) <- systems_list$system
# For each system, see if other systems are annotated to plasmids found in it: 
for (system in systems_list$system) {
  # Check the systems that this one system shares plasmids with 
  similar_systems <- plasmid_systems[grepl(pattern = paste0("\\b", system, "\\b"), x = plasmid_systems$systems, fixed = F), "systems"] %>% 
    strsplit(split = "|", fixed = T) %>% unlist() %>% unique()
  # If none of them have a representative yet, select this one 
  if (all(is.na(systems_list[similar_systems, "representative"]))) {
    representative <- system
  } else {
    # If some have, check if they all have the same representative
    non_NA_reps <- systems_list[similar_systems, ] %>% filter(!is.na(representative)) %>% pull(representative) %>% unique()
    if (length(non_NA_reps) == 1) {
      # If so, just select that same representative
      representative <- non_NA_reps[1]
    } else {
      # If they have several representatives, merge them by selecting the first and changing the others
      representative <- non_NA_reps[1]
      systems_list[systems_list$representative %in% non_NA_reps, "representative"] <- representative 
    }
  }
  # Then put the selected representative for the similar systems
  systems_list[similar_systems, "representative"] <- representative
}

# Save the newly annotated systems for each plasmid 
plasmid_systems$system <- plasmid_systems$systems
for (row in 1:nrow(plasmid_systems)) {
  # For each plasmid, get the list of systems it's annotated in 
  systems <- plasmid_systems[row, "systems"] %>% strsplit(split = "|", fixed = T) %>% unlist() %>% unique()
  # if it's NA, skip, it will remain NA
  if (any(is.na(systems))) {
    next
  }
  # Else, get the list of representatives
  reps <- systems_list[systems, "representative"]
  if(length(unique(reps)) > 1) {
    # Ensure there's no errors 
    print(paste0("ERROR: Plasmid ", plasmid_systems[row, "Plasmid_ID"], " has multiple representatives."))
  } else {
    plasmid_systems[row, "system"] <- reps[1]
  }
}
rownames(plasmid_systems) <- plasmid_systems$Plasmid_ID


# Change plasmid systems names so they are PSS74 instead of PS94|PS734|PS843|etc.
rename_plasmid_systems <- data.frame(systems = unique(plasmid_info$systems)[!is.na(unique(plasmid_info$systems))])
rename_plasmid_systems$newsystems <- paste0("PSS", 1:nrow(rename_plasmid_systems))

plasmid_info <- plasmid_info %>% left_join(rename_plasmid_systems, by = "systems") %>% select(-systems) %>% 
  rename(systems = newsystems)
rownames(plasmid_info) <- plasmid_info$Plasmid_ID



taxa_info <- plasmid_systems %>% select(system, Plasmid_ID) %>% cbind(plasmid_info[rownames(plasmid_systems), ] %>% select(-Plasmid_ID))

TAX <- tax_table(as.matrix(taxa_info))


counts_matrix <- numreads %>% select(- pCode) %>% as.matrix.data.frame()
rownames(counts_matrix) <- numreads$pCode

coverage_matrix <- coverage %>% select(- pCode) %>% as.matrix.data.frame() 
rownames(coverage_matrix) <- coverage$pCode

# Require a minimum coverage to count the reads mapping to a plasmid; otherwise convert it to 0:
#counts_matrix[coverage_matrix < 50] <- 0

OTU <- otu_table(counts_matrix, taxa_are_rows = TRUE)

phycount_plasmids <- phyloseq(OTU, TAX, MET)

counts_matrix_with_fragments <- numreads_with_fragments %>% select(- pCode) %>% as.matrix.data.frame() 
rownames(counts_matrix_with_fragments) <- numreads_with_fragments$pCode

coverage_matrix_with_fragments <- coverage_with_fragments %>% select(- pCode) %>% as.matrix.data.frame() 
rownames(coverage_matrix_with_fragments) <- coverage_with_fragments$pCode

OTU_oc <- otu_table(counts_matrix_with_fragments, taxa_are_rows = TRUE)
TAX_oc <- tax_table(as.matrix(taxa_info[rownames(counts_matrix_with_fragments),]))
phycount_plasmids_with_fragments <- phyloseq(OTU_oc, TAX_oc, MET)

```




## RPKM
```{r}
rpkm_matrix <- rpkm(counts_matrix, gene.length = plasmid_info[rownames(counts_matrix), "Length"], log = F, 
                        lib.size = joint_metadata[colnames(counts_matrix), "n_reads"])
OTUrpkm <- otu_table(rpkm_matrix, taxa_are_rows = TRUE)
phyrpkm_plasmids <- phyloseq(OTUrpkm, TAX, MET)

rpkm_matrix_log <- rpkm(counts_matrix, gene.length = plasmid_info[rownames(counts_matrix), "Length"], log = T, 
                        lib.size = joint_metadata[colnames(counts_matrix), "n_reads"])
OTUrpkm_log <- otu_table(rpkm_matrix_log, taxa_are_rows = TRUE)
phyrpkm_plasmids_log <- phyloseq(OTUrpkm_log, TAX, MET)


# rpkm_matrix_with_fragments <- rpkm(counts_matrix_with_fragments, gene.length = plasmid_info[rownames(counts_matrix_with_fragments), "Length"],
#                                       log = F, lib.size = joint_metadata[colnames(counts_matrix_with_fragments), "n_reads"])
# OTU_oc_rpkm <- otu_table(rpkm_matrix_with_fragments, taxa_are_rows = TRUE)
# phyrpkm_plasmids_with_fragments <- phyloseq(OTU_oc_rpkm, TAX_oc, MET)
# 
# rpkm_matrix_log_with_fragments <- rpkm(counts_matrix_with_fragments, gene.length = plasmid_info[rownames(counts_matrix_with_fragments), "Length"],
#                                       log = T, lib.size = joint_metadata[colnames(counts_matrix_with_fragments), "n_reads"])
# OTU_oc_rpkm_log <- otu_table(rpkm_matrix_log_with_fragments, taxa_are_rows = TRUE)
# phyrpkm_plasmids_with_fragments_log <- phyloseq(OTU_oc_rpkm_log, TAX_oc, MET)

```

## Distance matrix
```{r}
# Calculate distance between samples with Bray-Curtis
D <- bcdist(rpkm_matrix %>% t())
dist_tree <- hclust(D, method = "complete")

# D_oc <- bcdist(rpkm_matrix_with_fragments %>% t())
# dist_tree_oc <- hclust(D_oc, method = "complete")
```

## Gene-level info 

```{r}
# read gene-level reads
gene_reads <- read.table(file = "BASE_DIR/PROFILING/gene_profiling.tsv", header = T, sep = "\t")

# then left join on a table with just the gene_caller_id and the accession
gene_otu <- gene_content %>% select(gene_callers_id, accession, partial, source.y, accession, function.) %>% 
  left_join(gene_reads %>% select(gene_callers_id, Plasmid_ID, start_pos, end_pos, direction,
                                  all_of(joint_metadata$SubjectID)), by = "gene_callers_id") %>%
  # Separate the accessions that are multiple accessions separated by !!! into different rows 
  separate_longer_delim(cols = c(accession, function.), delim = "!!!")
gene_otu$rand_id <- paste0("g", rownames(gene_otu))
rownames(gene_otu) <- gene_otu$rand_id

gene_otu_matrix <- gene_otu %>% select(all_of(joint_metadata$SubjectID))

gene_info <- gene_otu %>% select(rand_id, gene_callers_id, Plasmid_ID, accession, partial, source.y, function., start_pos, 
                                 end_pos, direction) %>% 
  mutate(gene_length = end_pos - start_pos)


gene_rpkm_matrix <- rpkm(gene_otu_matrix, gene.length = gene_info[rownames(gene_otu_matrix), "gene_length"], log = F, 
                        lib.size = joint_metadata[colnames(gene_otu_matrix), "n_reads"])


gene_rpkm_COG20 <- data.frame(gene_rpkm_matrix) %>% rownames_to_column(var = "rand_id") %>% left_join(gene_info, by = "rand_id") %>% 
  filter(source.y == "COG20_FUNCTION") %>% select(accession, all_of(joint_metadata$SubjectID)) %>% 
  # then group by accession and sum the genes with the same accession 
  group_by(accession) %>% summarise(across(.cols = all_of(joint_metadata$SubjectID), 
                                               .fns = list(sum), .names = "{.col}")) %>% data.frame()
rownames(gene_rpkm_COG20) <- gene_rpkm_COG20$accession
gene_rpkm_COG20 <- gene_rpkm_COG20 %>% select(-accession)

gene_rpkm_pfam <- data.frame(gene_rpkm_matrix) %>% rownames_to_column(var = "rand_id") %>% left_join(gene_info, by = "rand_id") %>% 
  filter(source.y == "Pfam") %>% select(accession, all_of(joint_metadata$SubjectID)) %>% 
  # then group by accession and sum the genes with the same accession 
  group_by(accession) %>% summarise(across(.cols = all_of(joint_metadata$SubjectID), 
                                               .fns = list(sum), .names = "{.col}"))%>% data.frame()
rownames(gene_rpkm_pfam) <- gene_rpkm_pfam$accession
gene_rpkm_pfam <- gene_rpkm_pfam %>% select(-accession)

COG20_otu <- otu_table(gene_rpkm_COG20, taxa_are_rows = TRUE)
Pfam_otu <- otu_table(gene_rpkm_pfam, taxa_are_rows = TRUE)


COG20_info <- gene_info %>% filter(source.y == "COG20_FUNCTION") %>% 
  # group by accession 
  group_by(accession) %>% summarise(gene_callers_ids = paste0(gene_callers_id, collapse = ","), 
                                    Plasmid_IDs = paste0(Plasmid_ID, collapse = ","),
                                    source.y = unique(source.y),
                                    function. = unique(function.), 
                                    partial_each = paste0(partial, collapse = ","),
                                    avg_length = case_when(all(partial == 0) ~ round(mean(gene_length)),
                                                           all(partial == 1) ~ max(gene_length),
                                                           TRUE ~ round(mean(gene_length[partial == 0]))) ) %>% data.frame()
rownames(COG20_info) <- COG20_info$accession


Pfam_info <- gene_info %>% filter(source.y == "Pfam") %>% 
  # group by accession 
  group_by(accession) %>% summarise(gene_callers_ids = paste0(gene_callers_id, collapse = ","), 
                                    Plasmid_IDs = paste0(Plasmid_ID, collapse = ","),
                                    source.y = unique(source.y),
                                    function. = unique(function.), 
                                    partial_each = paste0(partial, collapse = ","),
                                    avg_length = case_when(all(partial == 0) ~ round(mean(gene_length)),
                                                           all(partial == 1) ~ max(gene_length),
                                                           TRUE ~ round(mean(gene_length[partial == 0]))) ) %>% data.frame()
rownames(Pfam_info) <- Pfam_info$accession

TAX_COG <- tax_table(as.matrix(COG20_info))
TAX_pfam <- tax_table(as.matrix(Pfam_info))


phyrpkm_COG <- phyloseq(COG20_otu, TAX_COG, MET)
phyrpkm_pfam <- phyloseq(Pfam_otu, TAX_pfam, MET)

```



## Tax glom for system and for rep types, etc

```{r}
glom_plasmids <- function(phy, group_variable, na_value = "none"){
  # Get a reordered taxa table so the grouping variable is to the left
  new_tax_table <- tax_table(phy) %>% data.frame() %>% select(all_of(group_variable), Plasmid_ID) %>% 
    replace_na(setNames(object = list(na_value), nm = group_variable))
  # Replace the tax table with the new one
  tax_table(phy) <- tax_table(as.matrix(new_tax_table))
  # do the tax glom 
  phy_glom <- tax_glom(phy, taxrank = group_variable)
  # Rename the taxa names so they are at the new taxa level 
  taxa_names(phy_glom) <- tax_table(phy_glom)[taxa_names(phy_glom), group_variable]
  return(phy_glom)
}


phy_system <- glom_plasmids(phyrpkm_plasmids, group_variable = "system")
phy_systems <- glom_plasmids(phyrpkm_plasmids, group_variable = "systems")
phy_rep <- glom_plasmids(phyrpkm_plasmids, group_variable = "rep_type.s.")
phy_relaxase <- glom_plasmids(phyrpkm_plasmids, group_variable = "relaxase_type.s.")
phy_mpf <- glom_plasmids(phyrpkm_plasmids, group_variable = "mpf_type")
phy_orit <- glom_plasmids(phyrpkm_plasmids, group_variable = "orit_type.s.")
phy_mobility <- glom_plasmids(phyrpkm_plasmids, group_variable = "predicted_mobility")
phy_cluster_type <- glom_plasmids(phyrpkm_plasmids, group_variable = "cluster_type")
phy_circularity <- glom_plasmids(phyrpkm_plasmids, group_variable = "Circularity", na_value = "linear")
phy_AMR <- glom_plasmids(phyrpkm_plasmids, group_variable = "ARG", na_value = "FALSE")


# phy_system_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "system")
# phy_systems_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "systems")
# phy_rep_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "rep_type.s.")
# phy_relaxase_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "relaxase_type.s.")
# phy_mpf_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "mpf_type")
# phy_orit_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "orit_type.s.")
# phy_mobility_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "predicted_mobility")
# phy_cluster_type_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "cluster_type")
# phy_circularity_oc <- glom_plasmids(phyrpkm_plasmids_with_fragments, group_variable = "Circularity", na_value = "linear")

```




## Phyloseq list
```{r}

phylist_plasmids <- list(phy_count = phycount_plasmids, phy_rpkm = phyrpkm_plasmids, phy_logrpkm = phyrpkm_plasmids_log,
                         coverage_matrix = coverage_matrix, annotation = gene_content, dist = D, dist_tree = dist_tree, 
                         phy_system_rpkm = phy_system, phy_systems_rpkm = phy_systems, phy_rep_rpkm = phy_rep, 
                         phy_relaxase_rpkm = phy_relaxase, phy_mpf_rpkm = phy_mpf, phy_orit_rpkm = phy_orit, 
                         phy_mobility_rpkm = phy_mobility, phy_cluster_type_rpkm = phy_cluster_type, 
                         phy_circularity_rpkm = phy_circularity, phy_AMR = phy_AMR,
                         phy_rpkm_COG20 = phyrpkm_COG, 
                         phy_rpkm_pfam = phyrpkm_pfam)

# phylist_plasmids_with_fragments <- list(phy_count = phycount_plasmids_with_fragments, phy_rpkm = phyrpkm_plasmids_with_fragments, 
#                                        phy_logrpkm = phyrpkm_plasmids_with_fragments_log,
#                                        coverage_matrix = coverage_matrix_with_fragments, 
#                                        annotation = gene_content %>% filter(Plasmid_ID %in% rownames(coverage_matrix_with_fragments)), 
#                                        dist = D_oc, dist_tree = dist_tree_oc,
#                                        phy_system_rpkm = phy_system_oc, phy_systems_rpkm = phy_systems_oc, phy_rep_rpkm = phy_rep_oc,
#                                        phy_relaxase_rpkm = phy_relaxase_oc, phy_mpf_rpkm = phy_mpf_oc, phy_orit_rpkm = phy_orit_oc, 
#                          phy_mobility_rpkm = phy_mobility_oc, phy_cluster_type_rpkm = phy_cluster_type, 
#                          phy_circularity_rpkm = phy_circularity)

## Add alpha diversity measures 
chao1_info <- chao1_from_physeq_abs(phy_obj = phylist_plasmids[["phy_count"]], variable = "Disease", palette = colors[["disease"]])$index
rownames(chao1_info) <- chao1_info$sam_name


shannon_info <- shannon_from_physeq(phy_obj = phylist_plasmids[["phy_rpkm"]], variable = "Disease", palette = colors[["disease"]])$index
rownames(shannon_info) <- shannon_info$sam_name

phylist_plasmids[["phy_count"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_count"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_count"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_count"]]@sam_data), "Chao1"]


phylist_plasmids[["phy_rpkm"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_logrpkm"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_logrpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_logrpkm"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_logrpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_systems_rpkm"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_systems_rpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_systems_rpkm"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_systems_rpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_rpkm_COG20"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_rpkm_COG20"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm_COG20"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_rpkm_COG20"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_rpkm_pfam"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_rpkm_pfam"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm_pfam"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_rpkm_pfam"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_AMR"]]@sam_data$Shannon_plasmids <- shannon_info[rownames(phylist_plasmids[["phy_AMR"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_AMR"]]@sam_data$Chao1_plasmids <- chao1_info[rownames(phylist_plasmids[["phy_AMR"]]@sam_data), "Chao1"]



## Bacterial diversity 
# Load bacterial taxonomy info 
luis_POP_bug_list <- read.csv(file = "/mnt/synology/DATA_ANALYSIS/LUISITO_IBD/luisito_POP_buglist_sp.tsv", header = T, 
                              sep = "\t", row.names = 1)
colnames(luis_POP_bug_list) <- gsub(pattern = ".", replacement = "_", x = colnames(luis_POP_bug_list), fixed = T)
colnames(luis_POP_bug_list) <- ifelse(grepl("^VH", colnames(luis_POP_bug_list)), paste0(colnames(luis_POP_bug_list), "B"),
                                      colnames(luis_POP_bug_list))

luis_POP_bact <- metaphlan4_to_phyloseq(metadata = phylist_plasmids[["phy_rpkm"]]@sam_data %>% data.frame() %>% 
                                          filter(Study != "Longitudinal Mycobiome", SubjectID != "HCPOP_0167"),
                                    tree = F, bug_list = luis_POP_bug_list, abund = 0, prev = 0)



Longitudinal_buglist <- read.csv(file = "/mnt/synology/ZIXUAN/LongitudinalMycobiome/OTU_PWY_TABLES/lm_bacteria_buglist_mergeBF.csv",
                                 header = T, sep = ",", row.names = 1) %>% 
  select(any_of(rownames(phylist_plasmids[["phy_rpkm"]]@sam_data))) %>%
  mutate_if(is.numeric, round)
longitudinal_bact <- metaphlan4_to_phyloseq(metadata = phylist_plasmids[["phy_rpkm"]]@sam_data %>% data.frame() %>% 
                                          filter(Study == "Longitudinal Mycobiome"),
                                    tree = F, bug_list = Longitudinal_buglist, abund = 0, prev = 0)

phy_bact <- merge_phyloseq(luis_POP_bact, longitudinal_bact)

phylist_plasmids[["phy_bact"]] <- phy_bact


## Add alpha diversity measures 
chao1_bact <- chao1_from_physeq_abs(phy_obj = phylist_plasmids[["phy_bact"]], variable = "Disease")$index
rownames(chao1_bact) <- chao1_bact$sam_name

shannon_bact <- shannon_from_physeq(phy_obj = phylist_plasmids[["phy_bact"]], variable = "Disease")$index
rownames(shannon_bact) <- shannon_bact$sam_name


phylist_plasmids[["phy_count"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_count"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_count"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_count"]]@sam_data), "Chao1"]

phylist_plasmids[["phy_rpkm"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_logrpkm"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_logrpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_logrpkm"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_logrpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_systems_rpkm"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_systems_rpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_systems_rpkm"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_systems_rpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_rpkm_COG20"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_rpkm_COG20"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm_COG20"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_rpkm_COG20"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_rpkm_pfam"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_rpkm_pfam"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm_pfam"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_rpkm_pfam"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_AMR"]]@sam_data$Shannon_bact <- shannon_bact[rownames(phylist_plasmids[["phy_AMR"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_AMR"]]@sam_data$Chao1_bact <- chao1_bact[rownames(phylist_plasmids[["phy_AMR"]]@sam_data), "Chao1"]





#Save phyloseq objects 
save(phylist_plasmids, colors, rename_plasmid_systems,
     file = "BASE_DIR/PROFILING/phyloseq_complete.RData")
```


## Load phyloseq
```{r}
load(file = "BASE_DIR/PROFILING/phyloseq_complete.RData")


```

Matching of cases with controls
```{r}
## Getting a subset of matched cases and controls 
sample_data <- phylist_plasmids[["phy_count"]]@sam_data %>% data.frame() %>% 
  mutate(casecontrol = ifelse(Disease == "Healthy", "Control", "Case"))
sample_data$Gender<- factor(sample_data$Gender, levels = unique(sample_data$Gender))

matching <- matchControls(formula = casecontrol ~ Age + Gender + BMI_value, data = sample_data, contlabel = "Control",
               caselabel = "Case", replace = FALSE)

keepIDs <- c(matching$cases, matching$controls)

eliminate_samples <- sample_data$SubjectID[!sample_data$SubjectID %in% keepIDs]
#Filter out samples from that are not in the matching 


phylist_plasmids[["phy_count"]] <- phylist_plasmids[["phy_count"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_rpkm"]] <- phylist_plasmids[["phy_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_logrpkm"]] <- phylist_plasmids[["phy_logrpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_system_rpkm"]] <- phylist_plasmids[["phy_system_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_systems_rpkm"]] <- phylist_plasmids[["phy_systems_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_rep_rpkm"]] <- phylist_plasmids[["phy_rep_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_relaxase_rpkm"]] <- phylist_plasmids[["phy_relaxase_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_mpf_rpkm"]] <- phylist_plasmids[["phy_mpf_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_orit_rpkm"]] <- phylist_plasmids[["phy_orit_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_mobility_rpkm"]] <- phylist_plasmids[["phy_mobility_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_cluster_type_rpkm"]] <- phylist_plasmids[["phy_cluster_type_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_circularity_rpkm"]] <- phylist_plasmids[["phy_circularity_rpkm"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_AMR"]] <- phylist_plasmids[["phy_AMR"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_rpkm_COG20"]] <- phylist_plasmids[["phy_rpkm_COG20"]] %>% subset_samples(!SubjectID %in% eliminate_samples)
phylist_plasmids[["phy_rpkm_pfam"]] <- phylist_plasmids[["phy_rpkm_pfam"]] %>% subset_samples(!SubjectID %in% eliminate_samples)

# Also filter out plasmids with not enough prevalence
filt_plasmids <- filter_prevalence_abundance_dataframe_no_loop(data.frame(phylist_plasmids[["phy_count"]]@otu_table), abundance = 0.001,
                                                               prevalence = 0)[[2]]
filt_systems <- filter_prevalence_abundance_dataframe_no_loop(data.frame(phylist_plasmids[["phy_systems_rpkm"]]@otu_table), abundance = 0.001,
                                                               prevalence = 0)[[2]]
filt_COG <- filter_prevalence_abundance_dataframe_no_loop(data.frame(phylist_plasmids[["phy_rpkm_COG20"]]@otu_table), abundance = 0.001,
                                                               prevalence = 0)[[2]]
filt_pfam <- filter_prevalence_abundance_dataframe_no_loop(data.frame(phylist_plasmids[["phy_rpkm_pfam"]]@otu_table), abundance = 0.001,
                                                               prevalence = 0)[[2]]
phylist_plasmids[["phy_count"]] <- phylist_plasmids[["phy_count"]] %>% subset_taxa(!Plasmid_ID %in% filt_plasmids)
phylist_plasmids[["phy_rpkm"]] <- phylist_plasmids[["phy_rpkm"]] %>% subset_taxa(!Plasmid_ID %in% filt_plasmids)
phylist_plasmids[["phy_logrpkm"]] <- phylist_plasmids[["phy_logrpkm"]] %>% subset_taxa(!Plasmid_ID %in% filt_plasmids)
phylist_plasmids[["phy_systems_rpkm"]] <- phylist_plasmids[["phy_systems_rpkm"]] %>% subset_taxa(!systems %in% filt_systems)
phylist_plasmids[["phy_rpkm_COG20"]] <- phylist_plasmids[["phy_rpkm_COG20"]] %>% subset_taxa(!accession %in% filt_COG)
phylist_plasmids[["phy_rpkm_pfam"]] <- phylist_plasmids[["phy_rpkm_pfam"]] %>% subset_taxa(!accession %in% filt_pfam)


# Recalculate distance matrices with Bray-Curtis
phylist_plasmids[["dist"]] <- bcdist(as.matrix(data.frame(otu_table(phylist_plasmids[["phy_rpkm"]]))) %>% t())
phylist_plasmids[["dist_tree"]] <- hclust(phylist_plasmids[["dist"]], method = "complete")

# phylist_plasmids_with_fragments[["dist"]] <- bcdist(as.matrix(data.frame(otu_table(phylist_plasmids_with_fragments[["phy_rpkm"]]))) %>% t())
# phylist_plasmids_with_fragments[["dist_tree"]] <- hclust(phylist_plasmids_with_fragments[["dist"]], method = "complete")

save(phylist_plasmids, colors, rename_plasmid_systems,
     file = "BASE_DIR/matched_phyloseq.RData")

```

# Load info

```{r}
load(file = "BASE_DIR/matched_phyloseq.RData")
```


# Exploratory plots

## Samples
```{r}

boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Study", vary = "n_reads", palette = viridis(n = 3))

boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Study", vary = "Weight", palette = viridis(n = 3))

boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Study", vary = "Height", palette = viridis(n = 3))

boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Study", vary = "Age", palette = viridis(n = 3))

boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Study", vary = "BMI_value", palette = viridis(n = 3))


ggplot(data.frame(sample_data(phylist_plasmids$phy_count)), 
       mapping = aes(x = Study, fill = factor(BMI, levels = c("obesity", "overweight", "healthy weight", "underweight")))) +
  geom_bar(position = "fill") + theme_minimal() +
  scale_fill_viridis(discrete = T, name = "BMI")

ggplot(data.frame(sample_data(phylist_plasmids$phy_count)), 
       mapping = aes(x = Disease, fill = factor(BMI, levels = c("obesity", "overweight", "healthy weight", "underweight")))) +
  geom_bar(position = "fill") + theme_minimal() +
  scale_fill_viridis(discrete = T, name = "BMI")

ggplot(data.frame(sample_data(phylist_plasmids$phy_count)), 
       mapping = aes(x = Study, fill = Gender)) +
  geom_bar(position = "fill") + theme_minimal() + 
  scale_fill_manual(values = colors[["gender"]])


```

## Plasmid quantity

```{r}
summed_plasmids <- otu_table(phylist_plasmids[["phy_rpkm"]]) %>% data.frame() %>% as.matrix() %>% apply(MARGIN = 2, FUN = sum)

phylist_plasmids[["phy_rpkm"]]@sam_data$summed_plasmid_rpkm <- summed_plasmids[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data)]

boxplot_from_physeq(phylist_plasmids[["phy_rpkm"]], varx = "Disease", vary = "summed_plasmid_rpkm", palette = viridis(n = 3))


# summed_plasmids_oc <- otu_table(phylist_plasmids_with_fragments[["phy_rpkm"]]) %>% data.frame() %>% as.matrix() %>% apply(MARGIN = 2, FUN = sum)
# 
# phylist_plasmids_with_fragments[["phy_rpkm"]]@sam_data$summed_plasmid_rpkm <- 
#   summed_plasmids_oc[rownames(phylist_plasmids_with_fragments[["phy_rpkm"]]@sam_data)]
# 
# boxplot_from_physeq(phylist_plasmids_with_fragments[["phy_rpkm"]], varx = "Disease", vary = "summed_plasmid_rpkm", palette = viridis(n = 3))
```


## Plasmids stats

```{r}
plasmids_data <- data.frame(tax_table(phylist_plasmids[["phy_count"]]))
plasmids_data$Length <- as.numeric(plasmids_data$Length)

ggplot(data = plasmids_data, mapping = aes(x = Length, fill = Circularity)) + geom_density(alpha = 0.5)  + theme_minimal() + 
  labs(title = "Length distribution") + scale_x_log10(labels = scales::comma) 
mean(plasmids_data$Length)
quantile(plasmids_data$Length)

ggplot(data = plasmids_data, mapping = aes(x = Length, fill = predicted_mobility)) + geom_density(alpha = 0.5)  + theme_minimal() + 
  labs(title = "Length distribution") + scale_x_log10(labels = scales::comma) 
table(plasmids_data$Circularity)
table(plasmids_data$predicted_mobility)
table(plasmids_data$ARG, plasmids_data$predicted_mobility)
table(plasmids_data$cluster_type)


ggplot(data = plasmids_data, mapping = aes(x = as.numeric(PlasX_Score), fill = predicted_mobility)) + geom_density(alpha = 0.5)  + theme_minimal() + 
  labs(title = "PlasX score distribution") + scale_x_log10(labels = scales::comma) 

ggplot(data = plasmids_data, mapping = aes(x = Length, fill = cluster_type)) + geom_density(alpha = 0.5)  + theme_minimal() + 
  labs(title = "Length distribution") + scale_x_log10(labels = scales::comma) 

```


# Heatmap
```{r}

heatmap_rpkm <- plot_rpkm_heatmap(phy = phylist_plasmids$phy_logrpkm, cluster_rows = phylist_plasmids$dist_tree, 
                  rowside_vars = c("Project" = "Study", "Disease", "Gender"), 
                  rowside_cols = list("Project" = colors$project,
                                      Disease = colors$disease,
                                      Gender = colors$gender),
                  # colside_vars = c("Circularity"),
                  # colside_cols = list(Circularity = colors$circularity),
                  row_title = paste0(nrow(sample_data(phylist_plasmids$phy_logrpkm)), " samples"), 
                  col_title = paste0(nrow(tax_table(phylist_plasmids$phy_logrpkm)), " plasmids"), row_dend_width = 0.1
                  )

heatmap_rpkm
```


# Coverage heatmap 
(detection)
```{r}


heatmap_cov <- plot_coverage_heatmap(phy = phylist_plasmids$phy_logrpkm,
                                  cov_matrix = phylist_plasmids$coverage_matrix,
                                  cluster_rows = phylist_plasmids$dist_tree, 
                  rowside_vars = c("Project" = "Study", "Disease", "Gender"), 
                  rowside_cols = list("Project" = colors$project,
                                      Disease = colors$disease,
                                      Gender = colors$gender),
                  # colside_vars = c("Circularity"),
                  # colside_cols = list(Circularity = colors$circularity),
                  row_title = paste0(nrow(sample_data(phylist_plasmids$phy_logrpkm)), " samples"), 
                  col_title = paste0(nrow(tax_table(phylist_plasmids$phy_logrpkm)), " plasmids"), row_dend_width = 0.1
                  )

heatmap_cov_oc <- plot_coverage_heatmap(phy = phylist_plasmids_with_fragments$phy_logrpkm,
                                         cov_matrix = phylist_plasmids_with_fragments$coverage_matrix, 
                                         cluster_rows = phylist_plasmids_with_fragments$dist_tree,
                  rowside_vars = c("Project" = "Study", "Disease", "Gender"), 
                  rowside_cols = list("Project" = colors$project,
                                      Disease = colors$disease,
                                      Gender = colors$gender),
                  # colside_vars = c("Circularity"),
                  # colside_cols = list(Circularity = colors$circularity),
                  row_title = paste0(nrow(sample_data(phylist_plasmids_with_fragments$phy_logrpkm)), " samples"), 
                  col_title = paste0(nrow(tax_table(phylist_plasmids_with_fragments$phy_logrpkm)), " plasmids"), row_dend_width = 0.1
                  )

heatmap_cov
heatmap_cov_oc

```

# Rel abundance

## All plasmids

Relative abundance of plasmid systems/replicases/etc
```{r}

# At the plasmid level: 
taxa_plot_list <- plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_rpkm"]], facet_condition = "Disease", top_func = sum, clean_names = F,
                                      return_abundance = T,
                     taxonomic_rank = "Plasmid_ID", plotlegend = T, n = 30, tax_glom_done = T, palette = colors$general_palette_other) 
taxa_plot <- taxa_plot_list$plot + 
  guides(fill = guide_legend(ncol = 1)) 
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S1A.png", plot = taxa_plot, width = 2000, height = 1000, scale = 2.6, dpi = 300, units = "px")


# At the plasmid system level:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_system_rpkm"]], facet_condition = "Disease", top_func = sum, drop_none = F,
                     taxonomic_rank = "system", plotlegend = T, n = 20, tax_glom_done = T, 
                     legend_label = "Plasmid system", palette = colors$general_palette_other)  + 
  guides(fill = guide_legend(ncol = 1)) 

# At the plasmid systems level:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_systems_rpkm"]], facet_condition = "Disease", top_func = sum, drop_none = F,
                     taxonomic_rank = "systems", plotlegend = T, n = 40, tax_glom_done = T, 
                     legend_label = "Plasmid systems", palette = colors$general_palette_other)  + 
  guides(fill = guide_legend(ncol = 1)) 

# Replicase:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_rep_rpkm"]], facet_condition = "Disease", top_func = sum, drop_none = F,
                         taxonomic_rank = "rep_type.s.", plotlegend = T, n = 40, tax_glom_done = T, 
                         legend_label = "Plasmid rep type", palette = colors$general_palette_other)  +
  guides(fill = guide_legend(ncol = 1)) + coord_cartesian(ylim = c(0, 0.15))

# Relaxase:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_relaxase_rpkm"]], facet_condition = "Disease", top_func = sum, drop_none = F,
                         taxonomic_rank = "relaxase_type.s.", plotlegend = T, n = 40, tax_glom_done = T, 
                         legend_label = "Plasmid relaxase type", palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) + coord_cartesian(ylim = c(0, 0.15))

# MPF:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_mpf_rpkm"]], facet_condition = "Disease", top_func = sum, drop_none = F,
                         taxonomic_rank = "mpf_type", plotlegend = T, n = 40, tax_glom_done = T, legend_label = "mpf type", 
                         palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) + coord_cartesian(ylim = c(0, 0.05))

# OriT:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_orit_rpkm"]], facet_condition = "Disease", top_func = sum, drop_none = F,
                         taxonomic_rank = "orit_type.s.", plotlegend = T, n = 40, tax_glom_done = T, legend_label = "OriT type", 
                         palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) + coord_cartesian(ylim = c(0, 0.05))

# Mobility:
mobility_plot <- plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_mobility_rpkm"]], facet_condition = "Disease", top_func = sum,
                         taxonomic_rank = "predicted_mobility", plotlegend = T, n = 40, tax_glom_done = T, 
                         legend_label = "Plasmid mobility", palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) 
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S1B.png", plot = mobility_plot, width = 2000, height = 1000, scale = 2.6, dpi = 300, units = "px")

# Cluster type:
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_cluster_type_rpkm"]], facet_condition = "Disease", top_func = sum,
                         taxonomic_rank = "cluster_type", plotlegend = T, n = 40, tax_glom_done = T, 
                         legend_label = "Plasmid cluster type", palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) 

# Circularity
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_circularity_rpkm"]], facet_condition = "Disease", top_func = sum,
                         taxonomic_rank = "Circularity", plotlegend = T, tax_glom_done = T, 
                         legend_label = "Plasmid circularity", palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) 


# AMR
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_AMR"]], facet_condition = "Disease", top_func = sum,
                         taxonomic_rank = "ARG", plotlegend = T, tax_glom_done = T, 
                         legend_label = "AMR", palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1))  + coord_cartesian(ylim = c(0, 0.05))

# Genes?
plot_taxa_abundances_rel(phy_obj = phylist_plasmids[["phy_rpkm_COG20"]], facet_condition = "Disease", top_func = sum,
                         taxonomic_rank = "function.", plotlegend = T, tax_glom_done = T, n = 25, non_unique = T,
                         legend_label = "COG20", palette = colors$general_palette)  +
  guides(fill = guide_legend(ncol = 1)) 
```

# Gene content
```{r}
phylist_plasmids[["annotation"]] %>% filter(Plasmid_ID %in% rownames(phylist_plasmids[["phy_count"]]@tax_table), source.y == "COG20_FUNCTION") %>% group_by(function.) %>% summarise(n = n()) %>% arrange(desc(n)) %>% head(n = 20)
```


# AMR

```{r}
resistance_rpkm <- (otu_table(phylist_plasmids[["phy_AMR"]]) %>% data.frame() %>% as.matrix())["TRUE",]

phylist_plasmids[["phy_rpkm"]]@sam_data$resistance_rpkm <- resistance_rpkm[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data)]

boxplot_from_physeq(phylist_plasmids[["phy_rpkm"]], varx = "Disease", vary = "resistance_rpkm", palette = viridis(n = 3))
```


# PcoA
```{r}

pcoa_results <- pcoa(phylist_plasmids[["dist"]])
pcoa_table <- cbind(sample_data(phylist_plasmids$phy_count),
                    as.data.frame(pcoa_results$vectors)[rownames(sample_data(phylist_plasmids$phy_count)),])
phylist_plasmids[["pcoa_table"]] <- pcoa_table
phylist_plasmids[["pcoa_results"]] <- pcoa_results

pcoa_results_oc <- pcoa(phylist_plasmids_with_fragments[["dist"]])
pcoa_table_oc <- cbind(sample_data(phylist_plasmids_with_fragments$phy_count),
                    as.data.frame(pcoa_results_oc$vectors)[rownames(sample_data(phylist_plasmids_with_fragments$phy_count)),])
phylist_plasmids_with_fragments[["pcoa_table"]] <- pcoa_table_oc
phylist_plasmids_with_fragments[["pcoa_results"]] <- pcoa_results_oc

```

Plots
```{r}

plot_pcoa(pcoa_table = phylist_plasmids[["pcoa_table"]], pcoa_results = phylist_plasmids[["pcoa_results"]], colorvar = "Disease", 
          shapevar = "Study", palette = colors[["disease"]], ellipse = T)

plot_pcoa(pcoa_table = phylist_plasmids_with_fragments[["pcoa_table"]], pcoa_results = phylist_plasmids_with_fragments[["pcoa_results"]], 
          colorvar = "Disease", shapevar = "Study", palette = colors[["disease"]], ellipse = T)


# Scree plots: 
scree_plot(phylist_plasmids[["pcoa_results"]], num = 20, color = colors[["only_complete"]]["all"])
scree_plot(phylist_plasmids_with_fragments[["pcoa_results"]], num = 20, color = colors[["only_complete"]]["only_complete"])

```

# UMAP
```{r}


plot_umap(phy_rpkm = phylist_plasmids[["phy_rpkm"]], n_neighbors = 20, colorvar = "Disease", shapevar = "Study",
          palette = colors[["disease"]], ellipse = T)

plot_umap(phy_rpkm = phylist_plasmids_with_fragments[["phy_rpkm"]], n_neighbors = 20, colorvar = "Disease", shapevar = "Study",
          palette = colors[["disease"]], ellipse = T)


```

# pBI143: HCPOP_0155_000000008383, HCPOP_0189_000000041952, HCPOP_235_000000008271, VH_CD_030B_000000000860
```{r}
pBI143_plasmids <- c('HCPOP_0007_000000036412','HCPOP_0033_000000008490','HCPOP_0066_000000047820','HCPOP_0088_000000010274','HCPOP_0097_000000006396','HCPOP_0102_000000017992','HCPOP_0116_000000037096','HCPOP_0119_000000010827','HCPOP_0129_000000026456','HCPOP_0155_000000008383','HCPOP_0159_000000002291','HCPOP_0165_000000000675','HCPOP_0189_000000041952','HCPOP_209_000000003020','HCPOP_235_000000008271','HCPOP_236_000000002929','HCPOP_249_000000032198','HCPOP_284_000000008845','HCPOP_314_000000001127','HCPOP_328_000000018647','HCPOP_334_000000006077','HCPOP_355_000000020009','HCPOP_356_000000007456','HCPOP_426_000000015176','HCPOP_452_000000010905','HCPOP_464_000000014772','HCPOP_480_000000026566','HCPOP_481_000000013547','HCPOP_492_000000007997','L2_7C_000000007442','L3_8C_000000020829','L5_3C_000000005828','L6_3C_000000023242','POP9_000000011225','VH_CD_030B_000000000860')

pBI143 <- otu_table(phylist_plasmids[["phy_count"]])[pBI143_plasmids, ] %>% data.frame() 
pBI143$Plasmid_ID <- rownames(pBI143)

pBI143 <- pivot_longer(pBI143, cols = all_of(sample_data(phylist_plasmids[["phy_rpkm"]])$SubjectID),
                       names_to = 'SubjectID', values_to = 'pBI143_reads') %>% 
  left_join(sample_data(phylist_plasmids[["phy_rpkm"]]), by = 'SubjectID') %>% 
  mutate(percent_pBI143 = pBI143_reads/n_reads * 100)

ggplot(pBI143, mapping = aes(x = Disease, y = log10(percent_pBI143 + 0.0000001), color = Disease)) +
  geom_boxplot() + theme_minimal() + facet_wrap(~ Plasmid_ID, scales = "free_y") +
  geom_point(position = "jitter", alpha = 0.3) +
  scale_color_manual(values = colors[["disease"]])

ggplot(pBI143, mapping = aes(x = Disease, y = log10(percent_pBI143 + 0.0000001), color = Disease)) +
  geom_violin() + theme_minimal() + facet_wrap(~ Plasmid_ID, scales = "free_y") +
  geom_point(position = "jitter", alpha = 0.2, size = 0.2) +
  scale_color_manual(values = colors[["disease"]])



pBI143 <- phylist_plasmids[["coverage_matrix"]][pBI143_plasmids, ] %>% data.frame() 
pBI143$Plasmid_ID <- rownames(pBI143)

pBI143 <- pivot_longer(pBI143, cols = all_of(sample_data(phylist_plasmids[["phy_rpkm"]])$SubjectID),
                       names_to = 'SubjectID', values_to = 'pBI143_cov') %>% 
  left_join(sample_data(phylist_plasmids[["phy_rpkm"]]), by = 'SubjectID')

ggplot(pBI143, mapping = aes(x = Disease, y = pBI143_cov, color = Disease)) +
  geom_boxplot() + theme_minimal() + facet_wrap(~ Plasmid_ID, scales = "free_y") +
  geom_point(position = "jitter", alpha = 0.3) +
  scale_color_manual(values = colors[["disease"]])

ggplot(pBI143, mapping = aes(x = Disease, y = pBI143_cov, color = Disease)) +
  geom_violin() + theme_minimal() + facet_wrap(~ Plasmid_ID, scales = "free_y") +
  geom_point(position = "jitter", alpha = 0.2, size = 0.2) +
  scale_color_manual(values = colors[["disease"]])

```


# Alpha diversity

## Chao1

```{r}
#Calculate again chao1 but with the subset of plasmids present in this cohort
## Add alpha diversity measures 
chao1_info <- chao1_from_physeq_abs(phy_obj = phylist_plasmids[["phy_count"]], variable = "Disease", palette = colors[["disease"]])$index
rownames(chao1_info) <- chao1_info$sam_name


shannon_info <- shannon_from_physeq(phy_obj = phylist_plasmids[["phy_rpkm"]], variable = "Disease", palette = colors[["disease"]])$index
rownames(shannon_info) <- shannon_info$sam_name

phylist_plasmids[["phy_count"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_count"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_count"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_count"]]@sam_data), "Chao1"]


phylist_plasmids[["phy_rpkm"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_rpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_logrpkm"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_logrpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_logrpkm"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_logrpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_systems_rpkm"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_systems_rpkm"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_systems_rpkm"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_systems_rpkm"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_rpkm_COG20"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_rpkm_COG20"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm_COG20"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_rpkm_COG20"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_rpkm_pfam"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_rpkm_pfam"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_rpkm_pfam"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_rpkm_pfam"]]@sam_data), "Chao1"]
phylist_plasmids[["phy_AMR"]]@sam_data$Shannon_plasmids_sub <- shannon_info[rownames(phylist_plasmids[["phy_AMR"]]@sam_data), "Shannon"]
phylist_plasmids[["phy_AMR"]]@sam_data$Chao1_plasmids_sub <- chao1_info[rownames(phylist_plasmids[["phy_AMR"]]@sam_data), "Chao1"]
```

```{r}
# chao1_results <- chao1_from_physeq_abs(phy_obj = phylist_plasmids[["phy_count"]], variable = "Disease", palette = colors[["disease"]])
# 
# chao1_results$plot

chao1_plasmid <- boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Chao1_plasmids", palette = colors[["disease"]]) +
  theme(legend.position = "none") + labs(title = "Chao1") + ylab("Plasmid Chao1")
boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Chao1_plasmids_sub", palette = colors[["disease"]])
chao1_bact <- boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Chao1_bact", palette = colors[["disease"]]) +
  theme(legend.position = "none") + labs(title = "Bacterial Chao1") + ylab("Bacterial Chao1")


 phylist_plasmids[["phy_count"]]@sam_data$Chao1_rel <- phylist_plasmids[["phy_count"]]@sam_data$Chao1_plasmids /
                                                            phylist_plasmids[["phy_count"]]@sam_data$Chao1_bact
chao1_rel <- boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Chao1_rel", palette = colors[["disease"]]) +
  theme(legend.position = "none") + labs(title = "Relative Chao1") + ylab("Plasmid Chao1 / Bacterial Chao1")

phylist_plasmids[["phy_count"]]@sam_data$Chao1_rel_sub <- phylist_plasmids[["phy_count"]]@sam_data$Chao1_plasmids_sub /
                                                            phylist_plasmids[["phy_count"]]@sam_data$Chao1_bact
boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Chao1_rel_sub", palette = colors[["disease"]])


ggsave(filename = "/home/arnau/Desktop/TFM Figures/2A1.png", plot = chao1_plasmid, width = 1000, height = 1000, scale = 1.7, dpi = 300, units = "px")
ggsave(filename = "/home/arnau/Desktop/TFM Figures/2B1.png", plot = chao1_rel, width = 1000, height = 1000, scale = 1.7, dpi = 300, units = "px")
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S2-1.png", plot = chao1_bact, width = 1000, height = 1000, scale = 1.7, dpi = 300, units = "px")

# chao1_results_systems <- chao1_from_physeq_relab(phy_obj = phylist_plasmids[["phy_systems_rpkm"]], variable = "Disease", palette = colors[["disease"]]) 
# chao1_results_systems$index %>% rename(SubjectID = sam_name) %>% left_join(y = phylist_plasmids[["phy_count"]]@sam_data %>% data.frame() %>% select(SubjectID, Chao1_bact, Disease), by = "SubjectID") %>% mutate(Chao1_rel = Chao1 / Chao1_bact ) %>% ggplot(aes(x = Disease, y = Chao1_rel, color = Disease)) + geom_boxplot()

```

## Shannon

```{r}
# shannon_results <- shannon_from_physeq(phy_obj = phylist_plasmids[["phy_rpkm"]], variable = "Disease", palette = colors[["disease"]])
# 
# shannon_results$plot

shannon_plasmid <- boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Shannon_plasmids", palette = colors[["disease"]]) +
  theme(legend.position = "none") + labs(title = "Shannon") + ylab("Plasmid Shannon")
boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Shannon_plasmids_sub", palette = colors[["disease"]])
shannon_bact <- boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Shannon_bact", palette = colors[["disease"]]) +
  theme(legend.position = "none") + labs(title = "Bacterial Shannon") + ylab("Bacterial Shannon")


phylist_plasmids[["phy_count"]]@sam_data$Shannon_rel <- phylist_plasmids[["phy_count"]]@sam_data$Shannon_plasmids -
                                                            phylist_plasmids[["phy_count"]]@sam_data$Shannon_bact
shannon_rel <- boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Shannon_rel", palette = colors[["disease"]]) +
  theme(legend.position = "none") + labs(title = "Relative Shannon") + ylab("Plasmid Shannon - Bacterial Shannon")

phylist_plasmids[["phy_count"]]@sam_data$Shannon_rel_sub <- phylist_plasmids[["phy_count"]]@sam_data$Shannon_plasmids_sub -
                                                            phylist_plasmids[["phy_count"]]@sam_data$Shannon_bact
boxplot_from_physeq(phylist_plasmids$phy_count, varx = "Disease", vary = "Shannon_rel_sub", palette = colors[["disease"]])


ggsave(filename = "/home/arnau/Desktop/TFM Figures/2A2.png", plot = shannon_plasmid, width = 1000, height = 1000, scale = 1.7, dpi = 300, units = "px")
ggsave(filename = "/home/arnau/Desktop/TFM Figures/2B2.png", plot = shannon_rel, width = 1000, height = 1000, scale = 1.7, dpi = 300, units = "px")
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S2-2.png", plot = shannon_bact, width = 1000, height = 1000, scale = 1.7, dpi = 300, units = "px")

# shannon_results_genes <- shannon_from_physeq(phy_obj = phylist_plasmids[["phy_rpkm_COG20"]], 
#                                           variable = "Disease", palette = colors[["disease"]])
# 
# shannon_results_genes$plot

Fig2 <- ggarrange(chao1_plasmid, shannon_plasmid, chao1_rel, shannon_rel)
ggsave(filename = "/home/arnau/Desktop/TFM Figures/2AB.png", plot = Fig2, width = 3000, height = 3000, scale = 1, dpi = 300, units = "px")

FigS2 <- ggarrange(chao1_bact, shannon_bact)
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S2.png", plot = FigS2, width = 3000, height = 1500, scale = 1, dpi = 300, units = "px")
```

# Beta diversity 

```{r}
beta_div <- plot_beta_div_plasmids(physeq = phylist_plasmids$phy_rpkm, dist = phylist_plasmids$dist, group = "Disease", 
                          pairwise_adonis = T, add_pval = T, colors = colors[["disease"]])


beta_div$plot

ggsave(filename = "/home/arnau/Desktop/TFM Figures/2C.png", plot = beta_div$plot, width = 3000, height = 1500, scale = 1, dpi = 300, units = "px")
```




# ANCOM-BC


```{r}
run_ancombc <- function(phyobject, taxa_level){
  phyobject@sam_data$Disease <- factor(phyobject@sam_data$Disease, levels = c("Healthy", "CD", "UC"))
  
  ancom_output <- ancombc2(
    data = phyobject, taxa_are_rows = T, assay_name = "RPKM", tax_level = taxa_level, rand_formula = NULL,
    meta_data = sample_data(phyobject), fix_formula = "Disease + Age + Gender + BMI_value",
    p_adj_method = "BH", pseudo_sens = T, prv_cut = 0.1, lib_cut = 0, s0_perc = 0.15, group = "Disease", struc_zero = T, 
    neg_lb = T, alpha = 0.05, n_cl = 44, verbose = TRUE, global = F, pairwise  = F, dunnet = FALSE, trend = FALSE, 
                    iter_control = list(tol = 0.01, max_iter = 20, 
                                        verbose = FALSE),
                    em_control = list(tol = 1e-5, max_iter = 100),
                    lme_control = lme4::lmerControl(),
                    mdfdr_control = list(fwer_ctrl_method = "holm", B = 100),
                    trend_control = list(contrast = NULL, node = NULL, solver = "ECOS", B = 100)
  )
  
  return(ancom_output)
}


ancom_output <- run_ancombc(phylist_plasmids[["phy_rpkm"]], taxa_level = "Plasmid_ID")
# ancom_output_oc <- run_ancombc(phylist_plasmids_with_fragments[["phy_rpkm"]], taxa_level = "Plasmid_ID")
ancom_output_system <- run_ancombc(phylist_plasmids[["phy_system_rpkm"]], taxa_level = "system")
ancom_output_systems <- run_ancombc(phylist_plasmids[["phy_systems_rpkm"]], taxa_level = "systems")
ancom_output_rep <- run_ancombc(phylist_plasmids[["phy_rep_rpkm"]], taxa_level = "rep_type.s.")
ancom_output_relaxase <- run_ancombc(phylist_plasmids[["phy_relaxase_rpkm"]], taxa_level = "relaxase_type.s.")
ancom_output_mpf <- run_ancombc(phylist_plasmids[["phy_mpf_rpkm"]], taxa_level = "mpf_type")
ancom_output_orit <- run_ancombc(phylist_plasmids[["phy_orit_rpkm"]], taxa_level = "orit_type.s.")
ancom_output_COG20 <- run_ancombc(phylist_plasmids[["phy_rpkm_COG20"]], taxa_level = "accession")
ancom_output_pfam <- run_ancombc(phylist_plasmids[["phy_rpkm_pfam"]], taxa_level = "accession")

save(ancom_output, ancom_output_system, ancom_output_systems, 
     ancom_output_rep, ancom_output_relaxase, ancom_output_mpf, ancom_output_orit, ancom_output_COG20, ancom_output_pfam,
     file = 'BASE_DIR/ancom_plasmids_matched_prova.RData')
```

## Load results
```{r}
load("BASE_DIR/ancom_plasmids_matched_prova.RData")
```



## Process

```{r}
process_ancombc_res <- function(ancom_output){
  res_prim <- ancom_output$res
  res_res <- res_prim %>% select(taxon, names(res_prim)[names(res_prim) %>% endsWith(c("CD","UC"))]) %>%
    #reshape the data so each lfc_CD and lfc_UC columns, etc. get put into 1 column (so two rows per taxon, one for CD and one for UC)
    pivot_longer(cols = -taxon, # keep taxon column as is
                 names_to = c(".value", "Disease"), # split column names into (column name) (Disease)
                 names_pattern = "(.*)_(DiseaseCD|DiseaseUC)" #extract name and disease with regex
                 ) %>%
    mutate(Disease = ifelse(Disease == "DiseaseCD", "CD", "UC")) %>% # Rename Disease values 
    data.frame()
  return(res_res)
}

process_struc_zeros <- function(ancom_output){
  res_struc_zeros <- ancom_output$zero_ind %>% rename(sz_healthy = `structural_zero (Disease = Healthy)`, 
                                                      sz_CD = `structural_zero (Disease = CD)`, sz_UC = `structural_zero (Disease = UC)`)
}

res_plasmids <- process_ancombc_res(ancom_output)
# res_plasmids_oc <- process_ancombc_res(ancom_output_oc)
res_system <- process_ancombc_res(ancom_output_system)
res_systems <- process_ancombc_res(ancom_output_systems)
res_rep <- process_ancombc_res(ancom_output_rep)
res_relaxase <- process_ancombc_res(ancom_output_relaxase)
res_mpf <- process_ancombc_res(ancom_output_mpf)
res_orit <- process_ancombc_res(ancom_output_orit)
res_COG20 <- process_ancombc_res(ancom_output_COG20)
res_pfam <- process_ancombc_res(ancom_output_pfam)

res_struc_zeros <- process_struc_zeros(ancom_output)
# res_struc_zeros_oc <- process_struc_zeros(ancom_output_oc)
res_struc_zeros_system <- process_struc_zeros(ancom_output_system)
res_struc_zeros_systems <- process_struc_zeros(ancom_output_systems)
res_struc_zeros_rep <- process_struc_zeros(ancom_output_rep)
res_struc_zeros_relaxase <- process_struc_zeros(ancom_output_relaxase)
res_struc_zeros_mpf <- process_struc_zeros(ancom_output_mpf)
res_struc_zeros_orit <- process_struc_zeros(ancom_output_orit)
res_struc_zeros_COG20 <- process_struc_zeros(ancom_output_COG20)
res_struc_zeros_pfam <- process_struc_zeros(ancom_output_pfam)
```

```{r}
write.table(x = res_plasmids %>% filter(diff == T, passed_ss == T, abs(lfc) > log(2)) %>% rename(Plasmid_ID = taxon) %>% 
              left_join(phylist_plasmids[["phy_rpkm"]]@tax_table %>% data.frame() %>% 
                          select(Plasmid_ID, Circularity, predicted_mobility, cluster_type, systems, 
                                 Resistance.Mechanism, Antibiotic, rep_type.s.) %>% 
                          unique, by = "Plasmid_ID") %>% 
              arrange(Disease, desc(lfc)), 
             file = "BASE_DIR/PLASMIDOME_results/Plasmids_ancomBC.tsv", sep = "\t", row.names = F)
```

### P-value dist
```{r}
p_value_distribution <- function(res_plasmids, disease, comparison){
  res_perm <- res_plasmids %>% filter(Disease == disease) %>%
    dplyr::transmute(taxon = taxon,
                     p = `p`,
                     ss_pass = `passed_ss`) %>%
  rowwise() %>%
  mutate(
      p_update = case_when(
          # For taxa with significant p-values that failed the sensitivity analysis,
          # we assign a random value uniformly drawn from the range [0.05, 1].
          p < 0.05 & ss_pass == FALSE ~ runif(1, min = 0.05, max = 1),
    TRUE ~ p 
  )) %>%
  ungroup()

  df_p = res_perm %>%
    dplyr::select(p, p_update) %>%
    pivot_longer(cols = p:p_update, names_to = "type", values_to = "value") 
  df_p$type = recode(df_p$type, 
                     `p` = "ANCOM-BC2 (Without Sensitivity Analysis)", 
                     `p_update` = "ANCOM-BC2 (With Sensitivity Analysis)")
  
  num_bins = 30
  fig_p = df_p %>%
    ggplot(aes(x = value, fill = type)) +
    geom_histogram(position = "identity", alpha = 0.5, bins = num_bins) +
    geom_hline(yintercept = nrow(res_perm)/num_bins, linetype = "dashed") +
    theme_minimal() +
    labs(title = paste0("Distribution of P-values (", comparison, " - ", disease, ")"),
         x = "P-value",
         y = "Frequency",
         fill = NULL,
         color = NULL) +
      theme_bw() + 
      theme(plot.title = element_text(hjust = 0.5),
            panel.grid.minor.y = element_blank(),
            legend.position = "bottom")
  return(fig_p)
}


pvaluedist_plasmidsCD <- p_value_distribution(res_plasmids, "CD", "plasmid")
pvaluedist_plasmidsUC <-p_value_distribution(res_plasmids, "UC", "plasmid")
# p_value_distribution(res_plasmids_oc, "CD", "complete plasmid")
# p_value_distribution(res_plasmids_oc, "UC", "complete plasmid")
p_value_distribution(res_system, "CD", "plasmid system")
p_value_distribution(res_system, "UC", "plasmid system")
p_value_distribution(res_systems, "CD", "plasmid systems")
p_value_distribution(res_systems, "UC", "plasmid systems")
p_value_distribution(res_rep, "CD", "rep")
p_value_distribution(res_rep, "UC", "rep")
p_value_distribution(res_relaxase, "CD", "relaxase")
p_value_distribution(res_relaxase, "UC", "relaxase")
p_value_distribution(res_mpf, "CD", "MPF")
p_value_distribution(res_mpf, "UC", "MPF")
p_value_distribution(res_orit, "CD", "oriT")
p_value_distribution(res_orit, "UC", "oriT")
pvaluedist_COGCD <- p_value_distribution(res_COG20, "CD", comparison = "COGs")
pvaluedist_COGUC <- p_value_distribution(res_COG20, "UC", comparison = "COGs")
pvaluedist_pfamCD <- p_value_distribution(res_pfam, "CD", comparison = "Pfams")
pvaluedist_pfamUC <- p_value_distribution(res_pfam, "UC", comparison = "Pfams")

pvaldistplot <- ggarrange(pvaluedist_plasmidsCD, pvaluedist_plasmidsUC, pvaluedist_COGCD, pvaluedist_COGUC, 
          pvaluedist_pfamCD, pvaluedist_pfamUC, common.legend = T, ncol = 2, nrow = 3)
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S5.png", plot = pvaldistplot, width = 2000, height = 3000, scale = 1.5, dpi = 300, units = "px")

```

## Volcano plot
```{r}
volcano_plot_ancombc <- function(res_ancombc, disease, title, colorspecial = NULL){
  res_ancombc %>% filter(Disease == disease) %>%
    mutate(color_var = case_when(q > 0.05 ~ "NS", passed_ss == F ~ "ss", 
                                 taxon %in% colorspecial ~ "AMR", # Just in cas I want to color AMR genes
                                 lfc > 0 ~ "Up", lfc < 0 ~ "Down"), 
           label = ifelse(q <= 0.05 & passed_ss == T, taxon, NA)) %>%
    ggplot(mapping = aes(x = lfc, y = -log10(q), color = color_var)) +
      geom_point(alpha = 0.75) + 
      geom_abline(slope = 0, intercept = -log10(0.05)) +
      geom_text_repel(aes(label = label), max.overlaps = 10, size = 3, na.rm = T) + 
      theme_minimal() + 
      scale_color_manual(values = c("Down" = "blue2", "Up" = "red1", "NS" = "#808080", "ss" = "#A0A0A0", "AMR" = "green1"), name = "") +
      xlab("log FC") + ggtitle(paste0(disease, " vs Healthy - ", title))
}

volcano_plot_ancombc(res_plasmids, "CD", "plasmid differential abundance", 
                     colorspecial = phylist_plasmids[["phy_rpkm"]]@tax_table %>% data.frame() %>% filter(!is.na(ARG)) %>% pull("Plasmid_ID"))
volcano_plot_ancombc(res_plasmids, "UC", "plasmid differential abundance", 
                     colorspecial = phylist_plasmids[["phy_rpkm"]]@tax_table %>% data.frame() %>% filter(!is.na(ARG)) %>% pull("Plasmid_ID"))
# volcano_plot_ancombc(res_plasmids_oc, "CD", "plasmid differential abundance (oc)")
# volcano_plot_ancombc(res_plasmids_oc, "UC", "plasmid differential abundance (oc)")
volcano_plot_ancombc(res_system, "CD", "plasmid system differential abundance")
volcano_plot_ancombc(res_system, "UC", "plasmid system differential abundance")
volcano_plot_ancombc(res_systems, "CD", "plasmid systems differential abundance")
volcano_plot_ancombc(res_systems, "UC", "plasmid systems differential abundance")
volcano_plot_ancombc(res_rep, "CD", "rep differential abundance")
volcano_plot_ancombc(res_rep, "UC", "rep differential abundance")
volcano_plot_ancombc(res_relaxase, "CD", "relaxase differential abundance")
volcano_plot_ancombc(res_relaxase, "UC", "relaxase differential abundance")
volcano_plot_ancombc(res_mpf, "CD", "mpf differential abundance")
volcano_plot_ancombc(res_mpf, "UC", "mpf differential abundance")
volcano_plot_ancombc(res_orit, "CD", "oriT differential abundance")
volcano_plot_ancombc(res_orit, "UC", "oriT differential abundance")
volcano_plot_ancombc(res_COG20, "CD", "COG20 differential abundance")
volcano_plot_ancombc(res_COG20, "UC", "COG20 differential abundance")
volcano_plot_ancombc(res_pfam, "CD", "Pfam differential abundance")
volcano_plot_ancombc(res_pfam, "UC", "Pfam differential abundance")

plot_COG_table <- res_COG20 %>% filter(Disease == "CD") %>%
    mutate(color_var = case_when(q > 0.05 ~ "NS", passed_ss == F ~ "NS", lfc > 0 ~ "Overabundant", lfc < 0 ~ "Underabundant")) %>% 
    left_join(phylist_plasmids[["phy_rpkm_COG20"]]@tax_table %>% data.frame() %>% select(accession, function.), 
              by = c("taxon" = "accession")) %>% 
  mutate(name = str_extract(function., "\\((\\w+)\\)")) %>% mutate(name = str_remove_all(name, "[()]")) 
rownames(plot_COG_table) <- plot_COG_table$taxon

plot_COG_table[duplicated(plot_COG_table$name), "name"] <- paste0(plot_COG_table[duplicated(plot_COG_table$name), "name"], 
                                                                  runif(n = length(plot_COG_table[duplicated(plot_COG_table$name), "name"]),
                                                                        min = 1, max = 999)) 
plot_COG_table[is.na(plot_COG_table$name), "name"] <- paste0(plot_COG_table[is.na(plot_COG_table$name), "name"], 
                                                                  runif(n = length(plot_COG_table[is.na(plot_COG_table$name), "name"]),
                                                                        min = 1, max = 999)) 
plot_COG_table["COG5852", "name"] <- "Small acid-soluble spore protein SspA/SspB/SspC/SspD"
plot_COG_table["COG4716", "name"] <- "Myosin-crossreactive antigen"
plot_COG_table["COG2929", "name"] <- NA
plot_COG_table["COG3335", "name"] <- "Transposase"
plot_COG_table["COG3976", "name"] <- NA
plot_COG_table["COG1149", "name"] <- "P-loop ATPase"
plot_COG_table["COG2014", "name"] <- NA
# plot_COG_table["", "name"] <- ""
# plot_COG_table["", "name"] <- ""
# plot_COG_table["", "name"] <- ""
# plot_COG_table["", "name"] <- ""

plot_COG_table <- plot_COG_table %>% mutate(label = ifelse(q <= 0.05 & passed_ss == T & abs(lfc) > log(2), name, NA))


COG_CD_PLOT <- plot_COG_table %>% ggplot(mapping = aes(x = lfc, y = -log10(q), color = color_var)) +
      geom_point(alpha = 0.75) + 
      geom_abline(slope = 0, intercept = -log10(0.05)) +
      geom_label_repel(aes(label = label), max.overlaps = 40, size = 3, na.rm = T, show.legend = F, alpha = 0.9) + 
      theme_minimal() + 
      scale_color_manual(values = c("Underabundant" = "#5C126EFF", "Overabundant" = "#E55C30FF", "NS" = "#808080"), name = "") +
      xlab("log FC") + ggtitle(paste0("CD", " vs Healthy - ", nrow(plot_COG_table %>% filter(q <= 0.05 & passed_ss == T, abs(lfc) >= log(2))), " differentially abundant COGs"))

ggsave(filename = "/home/arnau/Desktop/TFM Figures/4A.png", plot = COG_CD_PLOT, width = 3000, height = 2000, scale = 1, dpi = 300, units = "px")


plot_COG_table_UC <- res_COG20 %>% filter(Disease == "UC") %>%
    mutate(color_var = case_when(q > 0.05 ~ "NS", passed_ss == F ~ "NS", lfc > 0 ~ "Overabundant", lfc < 0 ~ "Underabundant")) %>% 
    left_join(phylist_plasmids[["phy_rpkm_COG20"]]@tax_table %>% data.frame() %>% select(accession, function.), 
              by = c("taxon" = "accession")) %>% 
  mutate(name = str_extract(function., "\\((\\w+)\\)")) %>% mutate(name = str_remove_all(name, "[()]")) 
rownames(plot_COG_table_UC) <- plot_COG_table_UC$taxon

plot_COG_table_UC[duplicated(plot_COG_table_UC$name), "name"] <- paste0(plot_COG_table_UC[duplicated(plot_COG_table_UC$name), "name"], 
                                                                  runif(n = length(plot_COG_table_UC[duplicated(plot_COG_table_UC$name), "name"]),
                                                                        min = 1, max = 999)) 
plot_COG_table_UC[is.na(plot_COG_table_UC$name), "name"] <- paste0(plot_COG_table_UC[is.na(plot_COG_table_UC$name), "name"], 
                                                                  runif(n = length(plot_COG_table_UC[is.na(plot_COG_table_UC$name), "name"]),
                                                                        min = 1, max = 999)) 

plot_COG_table_UC["COG3905", "name"] <- "PutA1"
plot_COG_table_UC["COG3886", "name"] <- "HKD family nuclease"
plot_COG_table_UC["COG5852", "name"] <- "SspA/SspB/SspC/SspD"
plot_COG_table_UC["COG4930", "name"] <- NA
plot_COG_table_UC["COG4908", "name"] <- NA
plot_COG_table_UC["COG2361", "name"] <- "MNT-HEPN toxin"
plot_COG_table_UC["COG4226", "name"] <- "HicB"
# plot_COG_table_UC["", "name"] <- ""
# plot_COG_table_UC["", "name"] <- ""




plot_COG_table_UC <- plot_COG_table_UC %>% mutate(label = ifelse(q <= 0.05 & passed_ss == T & abs(lfc) >= log(2), name, NA))


COG_UC_PLOT <- plot_COG_table_UC %>%
    ggplot(mapping = aes(x = lfc, y = -log10(q), color = color_var)) +
      geom_point(alpha = 0.75) + 
      geom_abline(slope = 0, intercept = -log10(0.05)) +
      geom_label_repel(aes(label = label), max.overlaps = 25, size = 3, na.rm = T, show.legend = F, alpha = 0.9, force = 1.4) + 
      theme_minimal() + 
      scale_color_manual(values = c("Underabundant" = "#5C126EFF", "Overabundant" = "#E55C30FF", "NS" = "#808080"), name = "") +
      xlab("log FC") + ggtitle(paste0("UC", " vs Healthy - ", nrow(plot_COG_table_UC %>% filter(q <= 0.05 & passed_ss == T & abs(lfc) >= log(2))), " differentially abundant COGs"))

ggsave(filename = "/home/arnau/Desktop/TFM Figures/4B.png", plot = COG_UC_PLOT, width = 3000, height = 2000, scale = 1, dpi = 300, units = "px")



Pfam_CD_plot <- res_pfam %>% filter(Disease == "CD") %>%
    mutate(color_var = case_when(q > 0.05 ~ "NS", passed_ss == F ~ "NS", lfc > 0 ~ "Overabundant", lfc < 0 ~ "Underabundant")) %>% 
    left_join(phylist_plasmids[["phy_rpkm_pfam"]]@tax_table %>% data.frame() %>% select(accession, function.), 
              by = c("taxon" = "accession")) %>%
    mutate(label = ifelse(q <= 0.05 & passed_ss == T & abs(lfc) >= log(2), function., NA)) %>%
    ggplot(mapping = aes(x = lfc, y = -log10(q), color = color_var)) +
      geom_point(alpha = 0.75) + 
      geom_abline(slope = 0, intercept = -log10(0.05)) +
      geom_label_repel(aes(label = label), max.overlaps = 30, size = 2, na.rm = T, show.legend = F, alpha = 0.9, force = 1) + 
      theme_minimal() + 
      scale_color_manual(values = c("Underabundant" = "#5C126EFF", "Overabundant" = "#E55C30FF", "NS" = "#808080"), name = "") +
      xlab("log FC") + ggtitle(paste0("CD", " vs Healthy - ", nrow(res_pfam %>% filter(Disease == "CD", q <= 0.05 & passed_ss == T & 
                                                                                         abs(lfc) >= log(2))),
                                      " differentially abundant Pfams"))
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S4A.png", plot = Pfam_CD_plot, width = 3000, height = 2000, scale = 1, dpi = 300, units = "px")

Pfam_UC_plot <- res_pfam %>% filter(Disease == "UC") %>%
    mutate(color_var = case_when(q > 0.05 ~ "NS", passed_ss == F ~ "NS", lfc > 0 ~ "Overabundant", lfc < 0 ~ "Underabundant")) %>% 
    left_join(phylist_plasmids[["phy_rpkm_pfam"]]@tax_table %>% data.frame() %>% select(accession, function.), 
              by = c("taxon" = "accession")) %>%
    mutate(label = ifelse(q <= 0.05 & passed_ss == T, function., NA)) %>%
    ggplot(mapping = aes(x = lfc, y = -log10(q), color = color_var)) +
      geom_point(alpha = 0.75) + 
      geom_abline(slope = 0, intercept = -log10(0.05)) +
      geom_label_repel(aes(label = label), max.overlaps = 30, size = 2, na.rm = T, show.legend = F, alpha = 0.9, force = 1) + 
      theme_minimal() + 
      scale_color_manual(values = c("Underabundant" = "#5C126EFF", "Overabundant" = "#E55C30FF", "NS" = "#808080"), name = "") +
      xlab("log FC") + ggtitle(paste0("UC", " vs Healthy - ", nrow(res_pfam %>% filter(Disease == "UC", q <= 0.05 & passed_ss == T & 
                                                                                         abs(lfc) >= log(2))),
                                      " differentially abundant Pfams"))
ggsave(filename = "/home/arnau/Desktop/TFM Figures/S4B.png", plot = Pfam_UC_plot, width = 3000, height = 2000, scale = 1, dpi = 300, units = "px")

```


## logFC heatmap
```{r}

fc_boxplot <- function(res_plasmids, res_struc_zeros, struc_0 = F, keep_failed = T){
  res_sig <- res_plasmids %>% filter(diff == TRUE) %>% 
    mutate(color = ifelse(passed_ss == TRUE, colors[["sensitivity"]]["passed"], colors[["sensitivity"]]["failed"]))
  
  if (struc_0) {
  # Add structural 0s data if desired
  res_sig_zeroes <- res_struc_zeros %>% pivot_longer(cols = c("sz_CD", "sz_UC"), names_to = "Disease", values_to = "sz_Disease", 
                                   names_prefix = "sz_") %>% filter(sz_healthy != sz_Disease) %>% data.frame() %>%
                                          # if 0 in Healthy, then it's positive FC in disease
    mutate(lfc = case_when(Disease == "CD" & sz_healthy == T ~ max(res_sig %>% filter(Disease == "CD") %>% pull(lfc)),
                           Disease == "CD" & sz_healthy == F ~ min(res_sig %>% filter(Disease == "CD") %>% pull(lfc)),
                           Disease == "UC" & sz_healthy == T ~ max(res_sig %>% filter(Disease == "UC") %>% pull(lfc)),
                           Disease == "UC" & sz_healthy == F ~ min(res_sig %>% filter(Disease == "UC") %>% pull(lfc))
                           ),
           se = NA, W = NA, p = 0, q = 0, diff = TRUE, passed_ss = TRUE, color = colors[["sensitivity"]]["struc_zero"]) %>% 
    select(all_of(names(res_sig)))
  
  res_sig <- rbind(res_sig, res_sig_zeroes)
  }
  if (!keep_failed){
    #remove results that failed the sensitivity analysis if specified
    res_sig <- res_sig %>% filter(passed_ss == T)
  }
  
  res_sig <- res_sig %>% arrange(lfc) %>% mutate(q_plot = round(q, digits = 3))
  res_sig$taxon <- factor(res_sig$taxon, levels = unique(res_sig$taxon))
  
  lo <- floor(min(res_sig$lfc))
  up <- ceiling(max(res_sig$lfc))
  mid <- 0
  fig <- res_sig %>% ggplot(aes(x = Disease, y = taxon, fill = lfc)) + 
    geom_tile() +
    scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
                         na.value = "white", midpoint = mid, limit = c(lo, up),
                         name = NULL) +
    geom_text(aes(Disease, taxon, label = q_plot, color = color), size = 4) +
      scale_color_identity(guide = "none") +
    labs(x = NULL, y = NULL, title = "Log fold changes as compared to healthy") +
    theme_minimal() +
    theme(plot.title = element_text(hjust = 0.5))
  
  return(fig)
}


fc_boxplot(res_plasmids, res_struc_zeros, struc_0 = F, keep_failed = F)
# fc_boxplot(res_plasmids_oc, res_struc_zeros_oc, struc_0 = F, keep_failed = F)
fc_boxplot(res_system, res_struc_zeros_system, struc_0 = F, keep_failed = F)
fc_boxplot(res_systems, res_struc_zeros_systems, struc_0 = F, keep_failed = F)
fc_boxplot(res_rep, res_struc_zeros_rep, struc_0 = F, keep_failed = F)
fc_boxplot(res_relaxase, res_struc_zeros_relaxase, struc_0 = F, keep_failed = F)
fc_boxplot(res_mpf, res_struc_zeros_mpf, struc_0 = F, keep_failed = F)
fc_boxplot(res_COG20, res_struc_zeros_COG20, struc_0 = F, keep_failed = F)

```




## Heatmap

Add metadata to IBD patients
```{r}
Luisito_metadata <- read_xlsx(path = 'BASE_DIR/SAMPLE_LISTS/metadata_IBD_Validation_CD_UC_Nov_24_GSG.xlsx',
                              sheet = 1, col_names = T) %>% 
  rename(Disease = `IBD type(1=CD, 2=UC)`) %>%
  mutate(SubjectID = gsub(pattern = '.', replacement = '_', x = paste0(SubjectID, 'B'), fixed = T)) %>% 
  filter(SubjectID %in% rownames(phylist_plasmids[["phy_count"]]@sam_data), Disease != '') %>% 
  mutate(Gender = case_when(gender == 'H' ~ 'm', gender == 'M' ~ 'f')) %>% 
  select(SubjectID, UC_Extension = `Extension UC: 1=Proctitis, 2=Izquierda, 3=extensa`, 
         CD_Phenotype = `Phenotype CD: 1=inflamatorio, 2=estenosante, 3= fistulizante + p (EPA)`,
         CD_Extension = `Extension CD: 1=ileal, 2=colnica, 3=ileocolnica, 4= GI superior`, 
         HB, CAI, 
         n_relapses = `number relapses...23`, 
         "MSZ","AZA","MTX","IFX","ADA" ,"GOLI","VDZ","USTK","TOFA","CORTICOSTEROIDS","Otro IMM", "Omeprazol","Antibiotics_last2months",
         Calprotectin = "Calprotectin (mg/kg)",
         C_Reactive_Prot = "C Reactive Protein (mg/dl)") %>% data.frame()
rownames(Luisito_metadata) <- Luisito_metadata$SubjectID
Luisito_metadata$Calprotectin <- as.numeric(Luisito_metadata$Calprotectin)
Luisito_metadata$CAI <- as.numeric(Luisito_metadata$CAI)
Luisito_metadata$C_Reactive_Prot <- as.numeric(Luisito_metadata$C_Reactive_Prot)


Luis_MET <- sample_data(Luisito_metadata)
phylist_plasmids[["phy_logrpkm"]] <- merge_phyloseq(phylist_plasmids[["phy_logrpkm"]], Luis_MET)
phylist_plasmids[["phy_rpkm"]] <- merge_phyloseq(phylist_plasmids[["phy_rpkm"]], Luis_MET)
phylist_plasmids[["phy_rpkm_COG20"]] <- merge_phyloseq(phylist_plasmids[["phy_rpkm_COG20"]], Luis_MET)


```



### Heatmap CD
```{r}

# cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "CD",
#                 legendname = "RPKM\nabundance", col_fontsize = 0, 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))))
# cd_hm$plot

cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "CD",
                legendname = "logRPKM\nabundance", col_fontsize = 0,
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))), 
                rowside_vars = c("CD_Phenotype", "CD_Extension" #,"MSZ", "AZA", "IFX", "ADA", "USTK", "CORTICOSTEROIDS", "Omeprazol", "Antibiotics_last2months"
                                 ), 
                rowside_cols = list(CD_Phenotype = c("B1" = "#FEC98DFF", "B2" = "#CD4071FF", "B2+p" = "#9F2F7FFF",
                                                     "B3" = "#451077FF", "B3+p" = "#180F3EFF", "NA" = "#bebebe"),
                                    CD_Extension = c("L1"= "#FEC98DFF", "L2" = "#CF4446FF", "L3" ="#781C6DFF", "L1+L4" = "#1B0C42FF", 
                                                     "NA" = "#bebebe")),  
                numeric_rowsidevars = c("HB", "Calprotectin", "C_Reactive_Prot", "n_relapses"), 
                numeric_rowside_cols = list("HB" = colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$HB,na.rm = T), 
                                                                mean(phylist_plasmids[["phy_logrpkm"]]@sam_data$HB,na.rm = T),
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$HB,na.rm = T)), 
                                                              c("#FEC98DFF", "#B41658FF", "#2A1636FF")),
                                            "Calprotectin"= colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$Calprotectin,na.rm = T), 
                                                                  150,
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$Calprotectin,na.rm = T)), 
                                                              c("#FEC98DFF", "#C4B56CFF", "#00204DFF")),
                                            "C_Reactive_Prot"= colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$C_Reactive_Prot,na.rm = T), 
                                                                mean(phylist_plasmids[["phy_logrpkm"]]@sam_data$C_Reactive_Prot,na.rm = T),
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$C_Reactive_Prot,na.rm = T)), 
                                                              c("#FEC98DFF", "#C4B56CFF", "#00336FFF")),
                                            "n_relapses"= colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$n_relapses,na.rm = T), 
                                                                mean(phylist_plasmids[["phy_logrpkm"]]@sam_data$n_relapses,na.rm = T),
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$n_relapses,na.rm = T)), 
                                                              c("#FEC98DFF", "#FB9A06FF", "#ED6925FF"))))
cd_hm$plot




# cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "CD",
#                 legendname = "RPKM\nabundance", fold_change_threshold = log(4), 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))))
# cd_hm$plot

cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "CD",
                legendname = "logRPKM\nabundance", fold_change_threshold = log(4), 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))))
cd_hm$plot



# Adding structural zeroes
# cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "CD",
#                 legendname = "RPKM\nabundance", col_fontsize = 0, 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
#                 struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
#                 only_struc_zeroes = F, n_struc_zeroes = 100)
# cd_hm$plot

cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "CD",
                legendname = "logRPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
                only_struc_zeroes = F, n_struc_zeroes = 100)
cd_hm$plot

# cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "CD",
#                 legendname = "RPKM\nabundance", fold_change_threshold = log(4), 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
#                 struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
#                 only_struc_zeroes = F, n_struc_zeroes = 100)
# cd_hm$plot

cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "CD",
                legendname = "logRPKM\nabundance", fold_change_threshold = log(4), 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
                only_struc_zeroes = F, n_struc_zeroes = 100)
cd_hm$plot


# Only structural zeroes

# cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "CD",
#                 legendname = "RPKM\nabundance", col_fontsize = 0, 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
#                 struc_zeros = res_struc_zeros, addtitle = "Just struc 0s - ",
#                 only_struc_zeroes = T, n_struc_zeroes = 200)
# cd_hm$plot

cd_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "CD",
                legendname = "logRPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, addtitle = "Just struc 0s - ",
                only_struc_zeroes = T, n_struc_zeroes = 200)
cd_hm$plot



cd_struc_zeroes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "CD",
                legendname = "logRPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, 
                only_struc_zeroes = T)
cd_struc_zeroes$plot

### ADD CLINICAL INFORMATION ON CD 

```

### Heatmap UC
```{r}

# UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "UC",
#                 legendname = "RPKM\nabundance", col_fontsize = 0, 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))))
# UC_hm$plot

UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "UC",
                legendname = "logRPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                rowside_vars = c("UC_Extension" #,"MSZ", "AZA", "IFX", "ADA", "USTK", "CORTICOSTEROIDS", "Omeprazol", "Antibiotics_last2months"
                                 ), 
                rowside_cols = list(UC_Extension = c("E1"= "#FEC98DFF", "E2" = "#CF4446FF", "E3" ="#380C6DFF", 
                                                     "NA" = "#bebebe")),  
                numeric_rowsidevars = c("CAI", "Calprotectin", "C_Reactive_Prot", "n_relapses"), 
                numeric_rowside_cols = list("CAI" = colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$CAI,na.rm = T), 
                                                                mean(phylist_plasmids[["phy_logrpkm"]]@sam_data$CAI,na.rm = T),
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$CAI,na.rm = T)), 
                                                              c("#FEC98DFF", "#B41658FF", "#2A1636FF")),
                                            "Calprotectin"= colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$Calprotectin,na.rm = T), 
                                                                  150,
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$Calprotectin,na.rm = T)), 
                                                              c("#FEC98DFF", "#C4B56CFF", "#00204DFF")),
                                            "C_Reactive_Prot"= colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$C_Reactive_Prot,na.rm = T), 
                                                                mean(phylist_plasmids[["phy_logrpkm"]]@sam_data$C_Reactive_Prot,na.rm = T),
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$C_Reactive_Prot,na.rm = T)), 
                                                              c("#FEC98DFF", "#C4B56CFF", "#00204DFF")),
                                            "n_relapses"= colorRamp2(c(min(phylist_plasmids[["phy_logrpkm"]]@sam_data$n_relapses,na.rm = T), 
                                                                mean(phylist_plasmids[["phy_logrpkm"]]@sam_data$n_relapses,na.rm = T),
                                                                max(phylist_plasmids[["phy_logrpkm"]]@sam_data$n_relapses,na.rm = T)), 
                                                              c("#FEC98DFF", "#FB9A06FF", "#ED6925FF"))))
UC_hm$plot


# UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "UC",
#                 legendname = "RPKM\nabundance", fold_change_threshold = log(4), 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))))
# UC_hm$plot
# 
# UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "UC",
#                 legendname = "logRPKM\nabundance", fold_change_threshold = log(4), 
#                 color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))))
# UC_hm$plot



# Adding structural zeroes
# UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "UC",
#                 legendname = "RPKM\nabundance", col_fontsize = 0, 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
#                 struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
#                 only_struc_zeroes = F, n_struc_zeroes = 100)
# UC_hm$plot

UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "UC",
                legendname = "logRPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
                only_struc_zeroes = F, n_struc_zeroes = 100)
UC_hm$plot


# UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "UC",
#                 legendname = "RPKM\nabundance", fold_change_threshold = log(4), 
#                 color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
#                 struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
#                 only_struc_zeroes = F, n_struc_zeroes = 100)
# UC_hm$plot
# 
# UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "UC",
#                 legendname = "logRPKM\nabundance", fold_change_threshold = log(4), 
#                 color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
#                 struc_zeros = res_struc_zeros, addtitle = "With struc 0s - ",
#                 only_struc_zeroes = F, n_struc_zeroes = 100)
# UC_hm$plot


# Only structural zeroes

UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm"]], ancombc_results = res_plasmids, disease = "UC",
                legendname = "RPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(0, 10, 15, 20, 25, 30, 40), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, addtitle = "Just struc 0s - ",
                only_struc_zeroes = T, n_struc_zeroes = 200)
UC_hm$plot

UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "UC",
                legendname = "logRPKM\nabundance", col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, addtitle = "Just struc 0s - ",
                only_struc_zeroes = T)
UC_hm$plot




UC_struczeroes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm"]], ancombc_results = res_plasmids, disease = "UC",
                legendname = "logRPKM\nabundance", fold_change_threshold = log(2), col_fontsize = 0, 
                color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))),
                struc_zeros = res_struc_zeros, 
                only_struc_zeroes = T)
UC_struczeroes$plot


```


### Systems
```{r}
CD_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_systems_rpkm"]], ancombc_results = res_systems, disease = "CD",
                legendname = "RPKM\nabundance", tax_level = "systems",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Plasmid Systems", struc_zeros = res_struc_zeros_systems, only_struc_zeroes = F, n_struc_zeroes = 10)
CD_hm$plot

UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_systems_rpkm"]], ancombc_results = res_systems, disease = "UC",
                legendname = "RPKM\nabundance", tax_level = "systems",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Plasmid Systems", struc_zeros = res_struc_zeros_systems, only_struc_zeroes = F, n_struc_zeroes = 10)
UC_hm$plot

```



### System
```{r}
CD_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_system_rpkm"]], ancombc_results = res_system, disease = "CD",
                legendname = "RPKM\nabundance", tax_level = "system",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Plasmid Systems", struc_zeros = res_struc_zeros_system, only_struc_zeroes = F, n_struc_zeroes = 10)
CD_hm$plot

UC_hm <- ancombc_heatmap(physeq = phylist_plasmids[["phy_system_rpkm"]], ancombc_results = res_system, disease = "UC",
                legendname = "RPKM\nabundance", tax_level = "system",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Plasmid Systems", struc_zeros = res_struc_zeros_system, only_struc_zeroes = F, n_struc_zeroes = 10)
UC_hm$plot

```


### Relaxase
```{r}
phylist_plasmids[["phy_relaxase_rpkm"]]@tax_table[,"relaxase_type.s."] <- gsub(pattern = ",", replacement = "_", 
                                            x = phylist_plasmids[["phy_relaxase_rpkm"]]@tax_table[,"relaxase_type.s."])
rownames(phylist_plasmids[["phy_relaxase_rpkm"]]@tax_table) <- gsub(pattern = ",", replacement = "_",
                                                                x = rownames(phylist_plasmids[["phy_relaxase_rpkm"]]@tax_table))
rownames(phylist_plasmids[["phy_relaxase_rpkm"]]@otu_table) <- gsub(pattern = ",", replacement = "_",
                                                                x = rownames(phylist_plasmids[["phy_relaxase_rpkm"]]@otu_table))
res_relaxase$taxon <- gsub(pattern = ",", replacement = "_",res_relaxase$taxon)
res_struc_zeros_relaxase$taxon <- gsub(pattern = ",", replacement = "_",res_struc_zeros_relaxase$taxon)

CD_rel <- ancombc_heatmap(physeq = phylist_plasmids[["phy_relaxase_rpkm"]], ancombc_results = res_relaxase, disease = "CD",
                legendname = "RPKM\nabundance", tax_level = "relaxase_type.s.",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Relaxase", struc_zeros = res_struc_zeros_relaxase, only_struc_zeroes = F, n_struc_zeroes = 10)
CD_rel$plot

UC_rel <- ancombc_heatmap(physeq = phylist_plasmids[["phy_relaxase_rpkm"]], ancombc_results = res_relaxase, disease = "UC",
                legendname = "RPKM\nabundance", tax_level = "relaxase_type.s.",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Relaxase", struc_zeros = res_struc_zeros_relaxase, only_struc_zeroes = F, n_struc_zeroes = 10)
UC_rel$plot

```

### Genes

#### COGs
```{r}
CD_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm_COG20"]], ancombc_results = res_COG20, disease = "CD",
                legendname = "RPKM\nabundance", tax_level = "accession", fold_change_threshold = 0,
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "COGs")
                #whatisdifferential = "COGs", struc_zeros = res_struc_zeros_COG20, only_struc_zeroes = F, n_struc_zeroes = 10)
CD_genes$plot

write.table(x = res_COG20 %>% filter(diff == T, passed_ss == T, Disease == "CD", abs(lfc) > log(2)) %>% rename(accession = taxon) %>% 
              left_join(phylist_plasmids[["phy_rpkm_COG20"]]@tax_table %>% data.frame() %>% 
                          select(accession, function.) %>% unique, by = "accession") %>% 
              arrange(desc(lfc)), 
            file = "BASE_DIR/PLASMIDOME_results/CD_COG_ancomBC.tsv", sep = "\t", row.names = F)

UC_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm_COG20"]], ancombc_results = res_COG20, disease = "UC",
                legendname = "RPKM\nabundance", tax_level = "accession",
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "COGs")
                #whatisdifferential = "COGs", struc_zeros = res_struc_zeros_COG20, only_struc_zeroes = F, n_struc_zeroes = 10)
UC_genes$plot

write.table(x = res_COG20 %>% filter(diff == T, passed_ss == T, Disease == "UC", abs(lfc) > log(2)) %>% rename(accession = taxon) %>% 
              left_join(phylist_plasmids[["phy_rpkm_COG20"]]@tax_table %>% data.frame() %>% 
                          select(accession, function.) %>% unique, by = "accession") %>% 
              arrange(desc(lfc)), 
            file = "BASE_DIR/PLASMIDOME_results/UC_COG_ancomBC.tsv", sep = "\t", row.names = F)

# CD_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm_COG20"]], ancombc_results = res_COG20, disease = "CD",
#                 legendname = "logRPKM\nabundance", tax_level = "accession",
#                 color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))), 
#                 whatisdifferential = "COGs")
#                 #whatisdifferential = "COGs", struc_zeros = res_struc_zeros_COG20, only_struc_zeroes = F, n_struc_zeroes = 100)
# CD_genes$plot
# 
# UC_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm_COG20"]], ancombc_results = res_COG20, disease = "UC",
#                 legendname = "logRPKM\nabundance", tax_level = "accession",
#                 color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))), 
#                 whatisdifferential = "COGs")
#                 #whatisdifferential = "COGs", struc_zeros = res_struc_zeros_COG20, only_struc_zeroes = F, n_struc_zeroes = 50)
# UC_genes$plot


# Extract significant gene orthologs
COG_CD_over <- CD_genes$FC %>% filter(logFC > 0) %>% pull(taxon)
COG_CD_under <- CD_genes$FC %>% filter(logFC < 0) %>% pull(taxon)

COG_UC_over <- UC_genes$FC %>% filter(logFC > 0) %>% pull(taxon)
COG_UC_under <- UC_genes$FC %>% filter(logFC < 0) %>% pull(taxon)

```

#### Pfams
```{r}
CD_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm_pfam"]], ancombc_results = res_pfam, disease = "CD",
                legendname = "RPKM\nabundance", tax_level = "accession", fold_change_threshold = log(4),
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Pfams")
#                whatisdifferential = "Pfams", struc_zeros = res_struc_zeros_pfam, only_struc_zeroes = F, n_struc_zeroes = 10)
CD_genes$plot

write.table(x = res_pfam %>% filter(diff == T, passed_ss == T, Disease == "CD", abs(lfc) > log(2)) %>% rename(accession = taxon) %>% 
              left_join(phylist_plasmids[["phy_rpkm_pfam"]]@tax_table %>% data.frame() %>% 
                          select(accession, function.) %>% unique, by = "accession") %>% 
              arrange(desc(lfc)), 
            file = "BASE_DIR/PLASMIDOME_results/CD_Pfam_ancomBC.tsv", sep = "\t", row.names = F)

UC_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_rpkm_pfam"]], ancombc_results = res_pfam, disease = "UC",
                legendname = "RPKM\nabundance", tax_level = "accession", fold_change_threshold = log(4),
                color_ramp = colorRamp2(c(0, 5, 10, 15, 20, 25, 30, 40, 50), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(8))), 
                whatisdifferential = "Pfams")
#                whatisdifferential = "Pfams", struc_zeros = res_struc_zeros_pfam, only_struc_zeroes = F, n_struc_zeroes = 10)
UC_genes$plot

write.table(x = res_pfam %>% filter(diff == T, passed_ss == T, Disease == "UC", abs(lfc) > log(2)) %>% rename(accession = taxon) %>% 
              left_join(phylist_plasmids[["phy_rpkm_pfam"]]@tax_table %>% data.frame() %>% 
                          select(accession, function.) %>% unique, by = "accession") %>% 
              arrange(desc(lfc)), 
            file = "BASE_DIR/PLASMIDOME_results/UC_Pfam_ancomBC.tsv", sep = "\t", row.names = F)

# CD_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm_pfam"]], ancombc_results = res_pfam, disease = "CD",
#                 legendname = "logRPKM\nabundance", tax_level = "accession",
#                 color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))), 
#                 whatisdifferential = "Pfams")
# #                whatisdifferential = "Pfams", struc_zeros = res_struc_zeros_pfam, only_struc_zeroes = F, n_struc_zeroes = 10)
# CD_genes$plot
# 
# UC_genes <- ancombc_heatmap(physeq = phylist_plasmids[["phy_logrpkm_pfam"]], ancombc_results = res_pfam, disease = "UC",
#                 legendname = "logRPKM\nabundance", tax_level = "accession",
#                 color_ramp = colorRamp2(c(-7, -5, -2, 0, 2, 5, 7), c("black", viridis_pal(begin = 0, end = 1, direction = 1)(6))), 
#                 whatisdifferential = "Pfams")
# #                whatisdifferential = "Pfams", struc_zeros = res_struc_zeros_pfam, only_struc_zeroes = F, n_struc_zeroes = 10)
# UC_genes$plot


# Extract significant gene orthologs
pfam_CD_over <- CD_genes$FC %>% filter(logFC > 0) %>% pull(taxon)
pfam_CD_under <- CD_genes$FC %>% filter(logFC < 0) %>% pull(taxon)

pfam_UC_over <- UC_genes$FC %>% filter(logFC > 0) %>% pull(taxon)
pfam_UC_under <- UC_genes$FC %>% filter(logFC < 0) %>% pull(taxon)

```


## COG Enrichment

### CD 
```{r}
#phylist_plasmids[["phy_count_COG20"]]@tax_table[COG_CD_over,"function."]


COG_category_CD_over <- enrichCOG(
  COG_CD_over,
  dtype = "category", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "CD", type = "category", abundance = "over")
 

COG_category_CD_under <- enrichCOG(
  COG_CD_under,
  dtype = "category", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "CD", type = "category", abundance = "under")

COG_pathway_CD_over <- enrichCOG(
  COG_CD_over,
  dtype = "pathway", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "CD", type = "pathway", abundance = "over")

COG_pathway_CD_under <- enrichCOG(
  COG_CD_under,
  dtype = "pathway", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "CD", type = "pathway", abundance = "under")

```

### UC

```{r}
COG_category_UC_over <- enrichCOG(
  COG_UC_over,
  dtype = "category", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "UC", type = "category", abundance = "over")

COG_category_UC_under <- enrichCOG(
  COG_UC_under,
  dtype = "category", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "UC", type = "category", abundance = "under")

# COG_pathway_UC_over <- enrichCOG(
#   COG_UC_over,
#   dtype = "pathway", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, minGSSize = 5,
#   #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
# )@result %>% mutate(Disease = "UC", type = "pathway", abundance = "over")

COG_pathway_UC_under <- enrichCOG(
  COG_UC_under,
  dtype = "pathway", pvalueCutoff = 0.05, pAdjustMethod = "BH", qvalueCutoff = 0.05, 
  #universe = phylist_plasmids[["annotation"]] %>% filter(source.y == "COG20_FUNCTION") %>% pull(accession) %>% unique()
)@result %>% mutate(Disease = "UC", type = "pathway", abundance = "under")

```


### Join
```{r}
COG_enrichment <- rbind(COG_category_CD_over, COG_category_CD_under, COG_pathway_CD_over, COG_pathway_CD_under,
                        COG_category_UC_over, COG_category_UC_under, COG_pathway_UC_under) %>% # and COG_pathway_UC_over
  filter(qvalue < 0.05) 
COG_enrichment[COG_enrichment$abundance == "over", "abundance"] <- "Overabundant"
COG_enrichment[COG_enrichment$abundance == "under", "abundance"] <- "Underabundant"

enrichmentplot <- ggplot(COG_enrichment, mapping = aes(x = Disease, y = Description, color = abundance, size = FoldEnrichment)) +
    geom_point() + theme_minimal() + scale_color_manual(values = c("Underabundant" = "#5C126EFF", "Overabundant" = "#E55C30FF"))

ggsave(filename = "/home/arnau/Desktop/TFM Figures/4C.png", plot = enrichmentplot, width = 3000, height = 2000, scale = 0.7, dpi = 300, units = "px")

```

# Clinical correlations
```{r}

CD_phy <- phylist_plasmids[["phy_rpkm"]] %>% subset_samples(Disease == "CD") %>% subset_taxa(!Plasmid_ID %in% 
      ((phylist_plasmids[["phy_rpkm"]] %>% subset_samples(Disease == "CD"))@otu_table %>% filter_prevalence_abundance_dataframe_no_loop(
        abundance = 0.01, prevalence = 0.5))[[2]])
corr_genes_diet <- corrplot_from_physeq_complexHeatmap_arnau(phy_obj = CD_phy, corrmethod = "spearman", 
                                                   interest_metadata = c("HB", "Calprotectin", "C_Reactive_Prot", "n_relapses"))
## No significant correlations 
corr_genes_diet$plot
```
```{r}

UC_phy <- phylist_plasmids[["phy_rpkm"]] %>% subset_samples(Disease == "UC") %>% subset_taxa(!Plasmid_ID %in% 
      ((phylist_plasmids[["phy_rpkm"]] %>% subset_samples(Disease == "UC"))@otu_table %>% filter_prevalence_abundance_dataframe_no_loop(
        abundance = 0.01, prevalence = 0.5))[[2]])
corr_genes_diet <- corrplot_from_physeq_complexHeatmap_arnau(phy_obj = UC_phy, corrmethod = "spearman", 
                                                   interest_metadata = c("CAI", "Calprotectin", "C_Reactive_Prot", "n_relapses"))
## No significant correlations 
corr_genes_diet$plot
```

# Shannon correlations
```{r}



correlation_from_physeq(phy_obj = phylist_plasmids[["phy_count"]] %>% subset_samples(Disease == "CD"), var1 = "Shannon_bact", var2 = "Shannon_plasmids")

correlation_from_physeq(phy_obj = phylist_plasmids[["phy_count"]] %>% subset_samples(Disease == "UC"), var1 = "Shannon_bact", var2 = "Shannon_plasmids")

correlation_from_physeq(phy_obj = phylist_plasmids[["phy_count"]] %>% subset_samples(Disease == "Healthy"), var1 = "Shannon_bact", var2 = "Shannon_plasmids")

phylist_plasmids[["phy_count"]]@sam_data %>% data.frame() %>% ggplot(aes(x = Shannon_bact, y = Shannon_plasmids, color = Disease)) + geom_point() + geom_smooth(method = "lm")


phylist_plasmids[["phy_count"]]@sam_data %>% data.frame() %>% pivot_longer(cols = c("Shannon_bact", "Shannon_plasmids", "Shannon_rel"), names_to = "Shannon_type", names_prefix = "Shannon_", values_to = "Shannon") %>% ggplot(aes(x = SubjectID, y = Shannon_type, fill = Shannon)) + 
  geom_tile()
```

